# Data Import

In [ ]:
!free -h

               total        used        free      shared  buff/cache   available
Mem:            12Gi       463Mi        10Gi       1.0Mi       2.1Gi        11Gi
Swap:             0B          0B          0B


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
from google.colab import auth
auth.authenticate_user()

In [ ]:
import pandas as pd
import numpy as np
from statistics import mean
import math

In [ ]:
user_data = pd.read_json('/content/drive/MyDrive/MOOCCube/entities/user.json', lines=True)
user_data

id  name                                       course_order  \
0        U_7001215   李喜锋  [C_course-v1:TsinghuaX+00740043_2x_2015_T2+sp,...   
1       U_10402446   五元香  [C_course-v1:TsinghuaX+00510888X+2019_T1, C_co...   
2       U_10359065    魏珊  [C_course-v1:TsinghuaX+00612642X+sp, C_course-...   
3        U_7423998   郭海滨  [C_course-v1:TsinghuaX+30240184_2X+sp, C_cours...   
4         U_545306   李其艳  [C_course-v1:TsinghuaX+20430064_2X+sp, C_cours...   
...            ...   ...                                                ...   
199194   U_9447602  羊舌半凡  [C_course-v1:TsinghuaX+00612642X+sp, C_course-...   
199195   U_7517918   焦彭越  [C_course-v1:TsinghuaX+20250064+sp, C_course-v...   
199196   U_8665537   锺才俊  [C_course-v1:TsinghuaX+30240243X+sp, C_course-...   
199197  U_10621245   忻盼曼  [C_course-v1:TsinghuaX+30700313X+2019_T1, C_co...   
199198   U_6212911    牵宣  [C_course-v1:TsinghuaX+10421094X_2015_2+sp, C_...   

                                              enroll_time  
0       [2017-05-01 11:07:53, 2017-05-17 10:07:17, 201...  
1              [2019-06-14 08:50:04, 2019-01-04 20:36:07]  
2              [2019-01-18 21:19:56, 2019-01-14 21:54:54]  
3       [2017-08-16 10:38:11, 2018-07-01 18:24:24, 201...  
4       [2018-09-05 15:40:40, 2019-02-28 10:08:49, 201...  
...                                                   ...  
199194  [2018-07-08 15:02:53, 2018-08-24 10:04:58, 201...  
199195         [2018-05-06 16:37:18, 2017-08-28 18:10:53]  
199196         [2018-03-12 11:02:22, 2018-05-03 10:29:23]  
199197  [2019-03-19 13:28:00, 2019-04-23 20:28:01, 201...  
199198  [2019-03-06 07:02:30, 2019-03-03 22:34:54, 201...  

[199199 rows x 4 columns]

Import main data which come from user entities. The historical sequence of course for each user will be the main input or data that will be used for modelling.

# Data Understanding and Preprocessing - Model Evaluation

In [ ]:
user_data.shape

(199199, 4)

In [ ]:
user_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199199 entries, 0 to 199198
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   id            199199 non-null  object
 1   name          199199 non-null  object
 2   course_order  199199 non-null  object
 3   enroll_time   199199 non-null  object
dtypes: object(4)
memory usage: 6.1+ MB


## Missing data

In [ ]:
user_data.isnull().sum()

id              0
name            0
course_order    0
enroll_time     0
dtype: int64

In [ ]:
user_data.isnull().sum().any()

False

In [ ]:
user_data.isnull().sum()

id              0
name            0
course_order    0
enroll_time     0
dtype: int64

In [ ]:
user_data.isnull().sum().any()

False

No missing data

## Check the number of courses taken for each student

In [ ]:
num_course = []

for i in range(len(user_data)):
  num_course.append(len(user_data['course_order'][i]))

user_data['number_of_course'] = num_course

In [ ]:
user_data

id  name                                       course_order  \
0        U_7001215   李喜锋  [C_course-v1:TsinghuaX+00740043_2x_2015_T2+sp,...   
1       U_10402446   五元香  [C_course-v1:TsinghuaX+00510888X+2019_T1, C_co...   
2       U_10359065    魏珊  [C_course-v1:TsinghuaX+00612642X+sp, C_course-...   
3        U_7423998   郭海滨  [C_course-v1:TsinghuaX+30240184_2X+sp, C_cours...   
4         U_545306   李其艳  [C_course-v1:TsinghuaX+20430064_2X+sp, C_cours...   
...            ...   ...                                                ...   
199194   U_9447602  羊舌半凡  [C_course-v1:TsinghuaX+00612642X+sp, C_course-...   
199195   U_7517918   焦彭越  [C_course-v1:TsinghuaX+20250064+sp, C_course-v...   
199196   U_8665537   锺才俊  [C_course-v1:TsinghuaX+30240243X+sp, C_course-...   
199197  U_10621245   忻盼曼  [C_course-v1:TsinghuaX+30700313X+2019_T1, C_co...   
199198   U_6212911    牵宣  [C_course-v1:TsinghuaX+10421094X_2015_2+sp, C_...   

                                              enroll_time  number_of_course  
0       [2017-05-01 11:07:53, 2017-05-17 10:07:17, 201...                 5  
1              [2019-06-14 08:50:04, 2019-01-04 20:36:07]                 2  
2              [2019-01-18 21:19:56, 2019-01-14 21:54:54]                 2  
3       [2017-08-16 10:38:11, 2018-07-01 18:24:24, 201...                 7  
4       [2018-09-05 15:40:40, 2019-02-28 10:08:49, 201...                10  
...                                                   ...               ...  
199194  [2018-07-08 15:02:53, 2018-08-24 10:04:58, 201...                 5  
199195         [2018-05-06 16:37:18, 2017-08-28 18:10:53]                 2  
199196         [2018-03-12 11:02:22, 2018-05-03 10:29:23]                 2  
199197  [2019-03-19 13:28:00, 2019-04-23 20:28:01, 201...                 5  
199198  [2019-03-06 07:02:30, 2019-03-03 22:34:54, 201...                 5  

[199199 rows x 5 columns]

## Filter Course Sequence

In [ ]:
user_data_range = user_data[(user_data["number_of_course"] >= 5) & (user_data["number_of_course"] <= 10) ]

Use appropriate minum and maximum number of courses for simplicity. I choose to have courses sequence within 5 to 10 courses for each student

In [ ]:
user_data_range

id  name                                       course_order  \
0        U_7001215   李喜锋  [C_course-v1:TsinghuaX+00740043_2x_2015_T2+sp,...   
3        U_7423998   郭海滨  [C_course-v1:TsinghuaX+30240184_2X+sp, C_cours...   
4         U_545306   李其艳  [C_course-v1:TsinghuaX+20430064_2X+sp, C_cours...   
7        U_7594103    巴鹏  [C_course-v1:TsinghuaX+30240184+sp, C_course-v...   
13       U_3234246   称语山  [C_course-v1:TsinghuaX+00740123_X+sp, C_course...   
...            ...   ...                                                ...   
199185   U_6380402   梁安康  [C_course-v1:XJTU+C00204+sp, C_course-v1:Tsing...   
199193   U_9447603   薛烨烨  [C_course-v1:TsinghuaX+64100033X+SP, C_course-...   
199194   U_9447602  羊舌半凡  [C_course-v1:TsinghuaX+00612642X+sp, C_course-...   
199197  U_10621245   忻盼曼  [C_course-v1:TsinghuaX+30700313X+2019_T1, C_co...   
199198   U_6212911    牵宣  [C_course-v1:TsinghuaX+10421094X_2015_2+sp, C_...   

                                              enroll_time  number_of_course  
0       [2017-05-01 11:07:53, 2017-05-17 10:07:17, 201...                 5  
3       [2017-08-16 10:38:11, 2018-07-01 18:24:24, 201...                 7  
4       [2018-09-05 15:40:40, 2019-02-28 10:08:49, 201...                10  
7       [2018-07-15 20:45:57, 2018-01-18 20:59:34, 201...                 6  
13      [2017-02-27 22:15:55, 2018-12-08 20:23:01, 201...                 5  
...                                                   ...               ...  
199185  [2018-09-19 16:51:11, 2017-09-20 11:11:01, 201...                 6  
199193  [2018-07-19 18:25:08, 2019-03-10 21:03:39, 201...                 8  
199194  [2018-07-08 15:02:53, 2018-08-24 10:04:58, 201...                 5  
199197  [2019-03-19 13:28:00, 2019-04-23 20:28:01, 201...                 5  
199198  [2019-03-06 07:02:30, 2019-03-03 22:34:54, 201...                 5  

[29870 rows x 5 columns]

## Data sample

In [ ]:
user_data_sample = user_data_range.sample(frac=0.3, random_state = 41)

In [ ]:
user_data_sample

id  name                                       course_order  \
6865     U_758099    漆玉  [C_course-v1:TsinghuaX+00740043X_2015_T2+sp, C...   
156830  U_9265115  太叔紫翠  [C_course-v1:TsinghuaX+10430484X_2015_2+sp, C_...   
57340   U_6915226   麦凝荷  [C_course-v1:TsinghuaX+00690212X+sp, C_course-...   
126720  U_2548979   赛寻双  [C_course-v1:TsinghuaX+10421075X_2015_2+sp, C_...   
93893   U_7150581   那妙竹  [C_course-v1:TsinghuaX+40050444X+sp, C_course-...   
...           ...   ...                                                ...   
58752   U_6635930   门元恺  [C_course-v1:TsinghuaX+Pr20170406-SC_p1+sp, C_...   
73942   U_3133278    声谷  [C_course-v1:TsinghuaX+30700313X+sp, C_course-...   
118769  U_1976347   似初南  [C_course-v1:CUNY+20171211001+sp, C_course-v1:...   
99802   U_5325210    陈承  [C_course-v1:TsinghuaX+70250023X_2015_2+sp, C_...   
161723  U_8028961   禽瀚漠  [C_course-v1:TsinghuaX+20220332X+sp, C_course-...   

                                              enroll_time  number_of_course  
6865    [2017-10-29 22:19:03, 2016-01-08 13:20:08, 201...                 9  
156830  [2018-06-14 18:08:34, 2018-09-09 20:12:24, 201...                 6  
57340   [2017-08-21 19:59:48, 2019-05-26 19:59:59, 201...                 5  
126720  [2018-05-02 10:12:06, 2017-07-04 08:21:05, 201...                 8  
93893   [2017-08-13 19:21:01, 2018-03-16 09:45:49, 201...                 5  
...                                                   ...               ...  
58752   [2017-08-14 10:53:21, 2017-08-04 07:36:53, 201...                 8  
73942   [2019-08-12 15:56:34, 2019-04-20 20:24:58, 201...                 6  
118769  [2019-05-19 10:19:37, 2018-11-16 17:07:59, 201...                 6  
99802   [2016-11-10 15:26:17, 2017-03-23 22:54:06, 201...                 9  
161723  [2017-11-18 19:03:30, 2017-11-10 11:07:46, 201...                 5  

[8961 rows x 5 columns]

Data is further filtered by sampling, only take 30% from 30K rows.

In [ ]:
def v_counts(dataframe):
    for i in dataframe :
        print(dataframe[i].value_counts())
        print("_____________________________________________________________________________")

v_counts(user_data_sample)

U_758099      1
U_9005717     1
U_9589535     1
U_8386334     1
U_101472      1
             ..
U_10443694    1
U_9889935     1
U_160094      1
U_459021      1
U_8028961     1
Name: id, Length: 8961, dtype: int64
_____________________________________________________________________________
漆玉     1
丁永桂    1
空锐立    1
厚天空    1
融鸿宝    1
      ..
庚欣嘉    1
仙信厚    1
陈令山    1
英凝雁    1
禽瀚漠    1
Name: name, Length: 8961, dtype: int64
_____________________________________________________________________________
[C_course-v1:MITx+6_00_1x+sp, C_course-v1:TsinghuaX+00740123_X+sp, C_course-v1:TsinghuaX+30240243X+sp, C_course-v1:TsinghuaX+30240184+sp, C_course-v1:TsinghuaX+30240184_2X+sp]                                                                                                                                                                                                              2
[C_course-v1:TsinghuaX+20250103X+sp, C_course-v1:TsinghuaX+20250064X+sp, C_course-v1:TsinghuaX+20250064+sp, C_

## Course Sequence Gathering and Transfomration

In [ ]:
course_order_list = user_data_sample["course_order"].tolist()
course_order_list

obtain the course sequence for each student and transform to list for further processing.

## Zero Padding

In [ ]:
zero_pad_course_order = []

for i in course_order_list:
  zero_pad_course_order.append([0]*(10-len(i))+i)

In [ ]:
zero_pad_course_order

Zero left pad to ensure all the data have similar length

## Check Sequence Length of the Data

In [ ]:
list_check = []
for i in zero_pad_course_order:
  list_check.append(len(i))

In [ ]:
array_check = np.array(list_check)
print(np.unique(array_check))

[10]


All data contained 10 sequence lengths.

## Course ID Gathering

### Extract all course ID in the dataset

In [ ]:
course_all_data = pd.read_json('/content/drive/MyDrive/MOOCCube/entities/course.json', lines=True)
course_all_data

id                   name  \
0             C_course-v1:McGillX+ATOC185x+2015_T1             自然灾害（自主模式）   
1        C_course-v1:TsinghuaX+THESIS2015X+2015_T1  2015年清华大学研究生学位论文答辩（二）   
2    C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_  2014年清华大学研究生学位论文答辩（一）   
3     C_course-v1:TsinghuaX+THESIS2015X_tv+2015_T1  2015年清华大学研究生学位论文答辩（一）   
4                C_course-v1:TsinghuaX+00690242+sp        文物精品与文化中国（自主模式）   
..                                             ...                    ...   
701             C_course-v1:TW+2018052501X+2019_T3   STEM课程设计与案例分析（2019秋）   
702        C_course-v1:TsinghuaX+60700052X+2019_T2          数据科学导论（2019夏）   
703        C_course-v1:TsinghuaX+80515182X+2019_sp         麦肯锡“全球领导力”自主模式   
704            C_course-v1:NEU+2019012201X+2019_T2         物理化学（上）（2019夏）   
705            C_course-v1:NEU+2019012202X+2019_T2         物理化学（下）（2019夏）   

                                  prerequisites  \
0                                             无   
1                                         无先修要求   
2                                         无先修要求   
3                                         无先修要求   
4                                             无   
..                                          ...   
701                                           无   
702                             本课程适合各类学科背景学生学习   
703  Basic knowledge of business administration   
704                                        高等数学   
705                                        高等数学   

                                                 about  \
0    <p>地球上没有一处地方不发生自然灾害。当我们以科学的眼光看待这些自然灾害的原因和本质时，我...   
1    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...   
2    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...   
3    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...   
4    <p>中国考古学是以往100年中发展最为迅速的领域之一，大批珍贵文物的出土，不断刷新人们对文...   
..                                                 ...   
701  <p>\r\n\t《STEM课程设计与案例分析》课程，它是：<br />\r\n1.国内领先...   
702  本课程作为数据科学的先导课和认知类课程，致力于以形象生动的教学模式为学生普及数据挖掘、大数据...   
703  <p>\r\n\t脱胎于清华经管学院与麦肯锡公司联合开设的线下金牌课程，麦肯锡全球领导力在线...   
704  物理化学（上）是一门运用物理学手段和技术研究化学问题的学科，被誉为化学学科的“大脑”。本课程...   
705  物理化学（下）是一门运用物理学手段和技术研究化学问题的学科，被誉为化学学科的“大脑”。本课程...   

                                           core_id  \
0             C_course-v1:McGillX+ATOC185x+2015_T1   
1        C_course-v1:TsinghuaX+THESIS2015X+2015_T1   
2    C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_   
3        C_course-v1:TsinghuaX+THESIS2015X+2015_T1   
4                C_course-v1:TsinghuaX+00690242+sp   
..                                             ...   
701             C_course-v1:TW+2018052501X+2019_T3   
702        C_course-v1:TsinghuaX+60700052X+2019_T1   
703        C_course-v1:TsinghuaX+80515182X+2019_sp   
704            C_course-v1:NEU+2019012201X+2019_T2   
705            C_course-v1:NEU+2019012202X+2019_T2   

                                           video_order  \
0    [V_f6f710068b994452885b90e11b6ee5c5, V_7339568...   
1    [V_de0371575a9f4b5391c89ad16d68b5c2, V_d632034...   
2    [V_d530be9cc0584317a16706684577a6dd, V_f329a62...   
3    [V_de0371575a9f4b5391c89ad16d68b5c2, V_78a8b41...   
4    [V_d7dbd0fe8f504e7a91d863cd5a19b185, V_4492eca...   
..                                                 ...   
701  [V_8f28c813854c444f98950e5404301f0b, V_f23a8ad...   
702  [V_2aa0936927744c3b9f1dcc8b5bc30fba, V_2def333...   
703  [V_ab97fe9d9a744958b886aa74d44146f8, V_8973856...   
704  [V_6595d9e0894848348287b8abaa30dce6, V_4dc4a13...   
705  [V_c21bcc55e0b84c539fb874cac8b24b68, V_8f1f9bb...   

                                          display_name  \
0    [Video: Overview 1, Video: Overview 2, Video: ...   
1    [答辩陈述, 答辩陈述, 问答及答辩结果, 答辩陈述, 问答及答辩结果, 答辩陈述, 问答及...   
2    [论文答辩实况, 问答及答辩结果, 导师评价, 同学眼中的王鑫, 个人学术感言, 吴宇恩答辩...   
3    [答辩陈述, 问答及答辩结果, 答辩陈述, 问答及答辩结果, 答辩陈述, 问答及答辩结果, ...   
4    [文献所见原始渡河工具, 《禹贡》所见的水路交通, 绰墩山出土的渡河浮木, 舟船考古发掘, ...   
..                                                 ...   
701  [何为 STEM 教育, STEM 教育的起源与发展, 世界各国的 STEM 教育政策, 为...   
702

In [ ]:
all_course = [i for i in course_all_data['id']]

In [ ]:
all_course

Gather all courses in the dataset.

### Extract all course ID in the sample data

In [ ]:
course_sample = [j for sub in course_order_list for j in sub]

In [ ]:
course_sample

In [ ]:
course_sample = np.unique(course_sample)

In [ ]:
course_sample

Gather all course in the data sample

In [ ]:
len(course_sample)

660

There are 660 courses in the data sample.


## Course and Teacher Relation

In [ ]:
teacher_course_relation = pd.read_csv('/content/drive/MyDrive/MOOCCube/relations/teacher-course.json', sep='\t', header=None)
teacher_course_relation

0                                          1
0     T_方维奇            C_course-v1:SPI+20170828001x+sp
1     T_方维奇      C_course-v1:SXPI+20170828001x+2019_T1
2     T_范茂魁      C_course-v1:PSFFC+2018102405X+2018_T2
3     T_连小珉  C_course-v1:TsinghuaX+70150104_2X+2019_T1
4     T_连小珉    C_course-v1:TsinghuaX+70150104X+2019_T1
...     ...                                        ...
2344  T_徐君莉        C_course-v1:NEU+2018122401X+2019_T1
2345   T_薛庆          C_course-v1:BIT+100070018+2019_T2
2346  T_史静琤          C_course-v1:CSU+20180919X+2019_T1
2347  T_刘小冰         C_course-v1:TsinghuaX+80515522X+sp
2348  T_黄湘晖        C_course-v1:XMU+2019031401X+2019_T1

[2349 rows x 2 columns]

In [ ]:
teacher_course_relation.set_axis(['teacher', 'course'], axis='columns', inplace=True)

<ipython-input-34-44f424829db4>:1: FutureWarning: DataFrame.set_axis 'inplace' keyword is deprecated and will be removed in a future version. Use `obj = obj.set_axis(..., copy=False)` instead
  teacher_course_relation.set_axis(['teacher', 'course'], axis='columns', inplace=True)


Set the columns name.

In [ ]:
teacher_course_relation

teacher                                     course
0      T_方维奇            C_course-v1:SPI+20170828001x+sp
1      T_方维奇      C_course-v1:SXPI+20170828001x+2019_T1
2      T_范茂魁      C_course-v1:PSFFC+2018102405X+2018_T2
3      T_连小珉  C_course-v1:TsinghuaX+70150104_2X+2019_T1
4      T_连小珉    C_course-v1:TsinghuaX+70150104X+2019_T1
...      ...                                        ...
2344   T_徐君莉        C_course-v1:NEU+2018122401X+2019_T1
2345    T_薛庆          C_course-v1:BIT+100070018+2019_T2
2346   T_史静琤          C_course-v1:CSU+20180919X+2019_T1
2347   T_刘小冰         C_course-v1:TsinghuaX+80515522X+sp
2348   T_黄湘晖        C_course-v1:XMU+2019031401X+2019_T1

[2349 rows x 2 columns]

In [ ]:
teacher_course_relation = teacher_course_relation.iloc[:,[1, 0]]

Change the position of each column.

In [ ]:
teacher_course_relation = teacher_course_relation[teacher_course_relation['course'].isin(course_sample)]

Filter the dataframe where it contains courses only from the sample data.

In [ ]:
teacher_course_relation

course teacher
0               C_course-v1:SPI+20170828001x+sp   T_方维奇
1         C_course-v1:SXPI+20170828001x+2019_T1   T_方维奇
2         C_course-v1:PSFFC+2018102405X+2018_T2   T_范茂魁
3     C_course-v1:TsinghuaX+70150104_2X+2019_T1   T_连小珉
4       C_course-v1:TsinghuaX+70150104X+2019_T1   T_连小珉
...                                         ...     ...
2344        C_course-v1:NEU+2018122401X+2019_T1   T_徐君莉
2345          C_course-v1:BIT+100070018+2019_T2    T_薛庆
2346          C_course-v1:CSU+20180919X+2019_T1   T_史静琤
2347         C_course-v1:TsinghuaX+80515522X+sp   T_刘小冰
2348        C_course-v1:XMU+2019031401X+2019_T1   T_黄湘晖

[2157 rows x 2 columns]

In [ ]:
len(np.unique(teacher_course_relation['course']))

651

A dataframe of course with their respective teacher is created. The dataframe consists only of courses from the sample data.

## Course and School Relation

In [ ]:
school_course_relation = pd.read_csv('/content/drive/MyDrive/MOOCCube/relations/school-course.json', sep='\t', header=None)
school_course_relation

0                                      1
0      S_BNU    C_course-v1:BNU+CSL21148501+2018_T2
1      S_BNU    C_course-v1:BNU+GE310141091+2019_T1
2      S_BNU    C_course-v1:BNU+2017112001X+2019_T1
3      S_BNU    C_course-v1:BNU+2018122603X+2018_T2
4      S_BNU    C_course-v1:BNU+CSL21126882+2019_T1
..       ...                                    ...
700  S_PSFFC  C_course-v1:PSFFC+2018102405X+2018_T2
701  S_PSFFC  C_course-v1:PSFFC+2018102403X+2018_T2
702  S_PSFFC  C_course-v1:PSFFC+2018102406X+2018_T2
703  S_PSFFC  C_course-v1:PSFFC+2018102404X+2018_T2
704    S_QLU    C_course-v1:QLU+2018122607X+2019_T1

[705 rows x 2 columns]

In [ ]:
school_course_relation.set_axis(['school', 'course'], axis='columns', inplace=True)

<ipython-input-41-ce3aad33b220>:1: FutureWarning: DataFrame.set_axis 'inplace' keyword is deprecated and will be removed in a future version. Use `obj = obj.set_axis(..., copy=False)` instead
  school_course_relation.set_axis(['school', 'course'], axis='columns', inplace=True)


Set the columns name.

In [ ]:
school_course_relation

school                                 course
0      S_BNU    C_course-v1:BNU+CSL21148501+2018_T2
1      S_BNU    C_course-v1:BNU+GE310141091+2019_T1
2      S_BNU    C_course-v1:BNU+2017112001X+2019_T1
3      S_BNU    C_course-v1:BNU+2018122603X+2018_T2
4      S_BNU    C_course-v1:BNU+CSL21126882+2019_T1
..       ...                                    ...
700  S_PSFFC  C_course-v1:PSFFC+2018102405X+2018_T2
701  S_PSFFC  C_course-v1:PSFFC+2018102403X+2018_T2
702  S_PSFFC  C_course-v1:PSFFC+2018102406X+2018_T2
703  S_PSFFC  C_course-v1:PSFFC+2018102404X+2018_T2
704    S_QLU    C_course-v1:QLU+2018122607X+2019_T1

[705 rows x 2 columns]

In [ ]:
school_course_relation = school_course_relation.iloc[:,[1, 0]]

Change the position of school and course feature.

In [ ]:
school_course_relation

course   school
0      C_course-v1:BNU+CSL21148501+2018_T2    S_BNU
1      C_course-v1:BNU+GE310141091+2019_T1    S_BNU
2      C_course-v1:BNU+2017112001X+2019_T1    S_BNU
3      C_course-v1:BNU+2018122603X+2018_T2    S_BNU
4      C_course-v1:BNU+CSL21126882+2019_T1    S_BNU
..                                     ...      ...
700  C_course-v1:PSFFC+2018102405X+2018_T2  S_PSFFC
701  C_course-v1:PSFFC+2018102403X+2018_T2  S_PSFFC
702  C_course-v1:PSFFC+2018102406X+2018_T2  S_PSFFC
703  C_course-v1:PSFFC+2018102404X+2018_T2  S_PSFFC
704    C_course-v1:QLU+2018122607X+2019_T1    S_QLU

[705 rows x 2 columns]

In [ ]:
school_course_relation = school_course_relation[school_course_relation['course'].isin(course_sample)]

Filter the dataframe where it contains courses only from the sample data.

In [ ]:
school_course_relation

course   school
0      C_course-v1:BNU+CSL21148501+2018_T2    S_BNU
1      C_course-v1:BNU+GE310141091+2019_T1    S_BNU
3      C_course-v1:BNU+2018122603X+2018_T2    S_BNU
4      C_course-v1:BNU+CSL21126882+2019_T1    S_BNU
5       C_course-v1:BNU+0610073991+2019_T1    S_BNU
..                                     ...      ...
699  C_course-v1:PSFFC+2018102402X+2018_T2  S_PSFFC
700  C_course-v1:PSFFC+2018102405X+2018_T2  S_PSFFC
701  C_course-v1:PSFFC+2018102403X+2018_T2  S_PSFFC
702  C_course-v1:PSFFC+2018102406X+2018_T2  S_PSFFC
704    C_course-v1:QLU+2018122607X+2019_T1    S_QLU

[659 rows x 2 columns]

A dataframe of course with their respective school is created. The dataframe consists only of courses from the sample data.

## Course Sequence Dataframe Construction

In [ ]:
course_columns = ['course_1', 'course_2', 'course_3', 'course_4', 'course_5', 'course_6', 'course_7', 'course_8', 'course_9', 'course_10']

Named the columns

In [ ]:
data_course = pd.DataFrame(zero_pad_course_order, columns = course_columns, dtype = str)

In [ ]:
data_course

course_1                                    course_2  \
0           0  C_course-v1:TsinghuaX+00740043X_2015_T2+sp   
1           0                                           0   
2           0                                           0   
3           0                                           0   
4           0                                           0   
...       ...                                         ...   
8956        0                                           0   
8957        0                                           0   
8958        0                                           0   
8959        0   C_course-v1:TsinghuaX+70250023X_2015_2+sp   
8960        0                                           0   

                                       course_3  \
0     C_course-v1:TsinghuaX+10421084X_2015_2+sp   
1                                             0   
2                                             0   
3     C_course-v1:TsinghuaX+10421075X_2015_2+sp   
4                                             0   
...                                         ...   
8956  C_course-v1:TsinghuaX+Pr20170406-SC_p1+sp   
8957                                          0   
8958                                          0   
8959                C_course-v1:WUT+1022817X+sp   
8960                                          0   

                                course_4  \
0     C_course-v1:TsinghuaX+00690092X+sp   
1                                      0   
2                                      0   
3     C_course-v1:TsinghuaX+20220332X+sp   
4                                      0   
...                                  ...   
8956  C_course-v1:TsinghuaX+20250103X+sp   
8957                                   0   
8958                                   0   
8959       C_course-v1:MITx+15_390_2x+sp   
8960                                   0   

                                       course_5  \
0            C_course-v1:TsinghuaX+30240243X+sp   
1     C_course-v1:TsinghuaX+10430484X_2015_2+sp   
2                                             0   
3       C_course-v1:TsinghuaX+80240372X+2019_T1   
4                                             0   
...                                         ...   
8956         C_course-v1:TsinghuaX+81020142X+sp   
8957         C_course-v1:TsinghuaX+30700313X+sp   
8958            C_course-v1:CUNY+20171211001+sp   
8959    C_course-v1:TsinghuaX+70660542X+2015_T2   
8960                                          0   

                                       course_6  \
0     C_course-v1:TsinghuaX+10430484X_2015_2+sp   
1     C_course-v1:TsinghuaX+20330334X_2015_2+sp   
2            C_course-v1:TsinghuaX+00690212X+sp   
3     C_course-v1:TsinghuaX+10421094X_2015_2+sp   
4            C_course-v1:TsinghuaX+40050444X+sp   
...                                         ...   
8956         C_course-v1:TsinghuaX+30260112X+sp   
8957    C_course-v1:TsinghuaX+AP000016X+2019_T1   
8958        C_course-v1:WageningenX+NUTR101x+sp   
8959         C_course-v1:TsinghuaX+00690342X+sp   
8960         C_course-v1:TsinghuaX+20220332X+sp   

                                       course_7  \
0          C_course-v1:TsinghuaX+20430064_2X+sp   
1     C_course-v1:TsinghuaX+10430494X_2015_2+sp   
2       C_course-v1:TsinghuaX+00670122X+2019_T1   
3                C_course-v1:RiceX+AdvBIO_1x+sp   
4                 C_course-v1:XIYOU+20180208+sp   
...                                         ...   
8956              C_course-v1:NJU+C1026+2016_T2   
8957              C_course-v1:MITx+15_390_2x+sp   
8958  C_course-v1:TsinghuaX+2018032801X+2018_T1   
8959         C_course-v1:TsinghuaX+60240202X+sp   
8960         C_course-v1:TsinghuaX+20250103X+sp   

                                        course_8  \
0      C_course-v1:TsinghuaX+10430494X_2015_2+sp   
1     C_course-v1:TsinghuaX+20220053X_2015_T2+sp   
2               C_course-v1:Tsinghua+20150001+sp   
3      C_course-v1:TsinghuaX+10430484X_2015_2+sp   
4              C_course-v1:Tsin

The course data consists of 10 columns from the first course until the final course order for each student (number or row represent number of stundent).

## School Sequence Dataframe Construction

In [ ]:
data_school = data_course.copy()

In [ ]:
dict_lookup_course_school = dict(zip(school_course_relation['course'], school_course_relation['school']))
dict_lookup_course_school

{'C_course-v1:BNU+CSL21148501+2018_T2': 'S_BNU',
 'C_course-v1:BNU+GE310141091+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018122603X+2018_T2': 'S_BNU',
 'C_course-v1:BNU+CSL21126882+2019_T1': 'S_BNU',
 'C_course-v1:BNU+0610073991+2019_T1': 'S_BNU',
 'C_course-v1:BNU+1010070372+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2017053101X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+GE410081071-01+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018122602X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091301X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091302X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091303X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+PHI2107404101+2019_T1': 'S_BNU',
 'C_course-v1:BNU+0610073981+2019_T1': 'S_BNU',
 'C_course-v1:BNU+0210021441+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2017071001X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091305X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018011801X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+ENV13018+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2017001+2018_T2': 'S_BNU',
 'C_course-v1:BNU+GOV21089

In [ ]:
dict_lookup_course_school['0'] = '0'

In [ ]:
len(dict_lookup_course_school)

660

In [ ]:
dict_lookup_course_school['C_course-v1:TsinghuaX+AP000003X+2019_T1'] = 'S_TsinghuaX'

Some of the courses did not have a school name. So add a school name for the course without the school name using the majority of school names in the data.

The dictionary lookup is constructed.

In [ ]:
for i in course_columns:
  data_school[i] = [dict_lookup_course_school[item] for item in data_school[i]]

Encode the course name with the school name.

In [ ]:
school_columns = ['school_1', 'school_2', 'school_3', 'school_4', 'school_5', 'school_6', 'school_7', 'school_8', 'school_9', 'school_10']

In [ ]:
data_school.columns = school_columns

Named the columns

In [ ]:
data_school

school_1     school_2     school_3     school_4     school_5  \
0           0  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX   
1           0            0            0            0  S_TsinghuaX   
2           0            0            0            0            0   
3           0            0  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX   
4           0            0            0            0            0   
...       ...          ...          ...          ...          ...   
8956        0            0  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX   
8957        0            0            0            0  S_TsinghuaX   
8958        0            0            0            0       S_CUNY   
8959        0  S_TsinghuaX        S_WUT       S_MITx  S_TsinghuaX   
8960        0            0            0            0            0   

           school_6     school_7     school_8     school_9    school_10  
0       S_TsinghuaX  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX  
1       S_TsinghuaX  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX        S_cqu  
2       S_TsinghuaX  S_TsinghuaX   S_TSINGHUA        S_UQx  S_TsinghuaX  
3       S_TsinghuaX      S_RiceX  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX  
4       S_TsinghuaX      S_XIYOU  S_TsinghuaX        S_UQx       S_SCUT  
...             ...          ...          ...          ...          ...  
8956    S_TsinghuaX        S_NJU  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX  
8957    S_TsinghuaX       S_MITx       S_MITx  S_TsinghuaX        S_UQx  
8958  S_WageningenX  S_TsinghuaX       S_MITx      S_RiceX  S_TsinghuaX  
8959    S_TsinghuaX  S_TsinghuaX     S_FUDANx       S_MITx  S_TsinghuaX  
8960    S_TsinghuaX  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX  S_TsinghuaX  

[8961 rows x 10 columns]

The school data consists of 10 columns from the first school until the final school order for each student (number or row represent number of stundent).

## Teacher Sequence Dataframe Construction

In [ ]:
data_teacher = data_course.copy()

In [ ]:
dict_lookup_course_teacher = dict(zip(teacher_course_relation['course'], teacher_course_relation['teacher']))
dict_lookup_course_teacher

{'C_course-v1:SPI+20170828001x+sp': 'T_姚常青',
 'C_course-v1:SXPI+20170828001x+2019_T1': 'T_姚常青',
 'C_course-v1:PSFFC+2018102405X+2018_T2': 'T_赵石楠',
 'C_course-v1:TsinghuaX+70150104_2X+2019_T1': 'T_连小珉',
 'C_course-v1:TsinghuaX+70150104X+2019_T1': 'T_连小珉',
 'C_course-v1:TJU+2010241X+sp': 'T_张德顺',
 'C_course-v1:TsinghuaX+20250103X+sp': 'T_任艳频',
 'C_course-v1:KMUSTX+8219011+2019_T1': 'T_王裕森',
 'C_course-v1:BIFT+2018122901X+2018_T2': 'T_丁雅琼',
 'C_course-v1:PSFFC+2018102403X+2018_T2': 'T_陶昆',
 'C_course-v1:BUCM+2018122604X+2018_T2': 'T_周芬',
 'C_course-v1:BUCM+2018122604X+2019_T1': 'T_周芬',
 'C_course-v1:TsinghuaX+20180919X+2019_T1': 'T_靳卫萍',
 'C_course-v1:SCUT+145033+sp': 'T_杨俊荣',
 'C_course-v1:BNU+2018091303X+2019_T1': 'T_孙璞',
 'C_course-v1:BNU+2018091304X+2019_T1': 'T_杨兆春',
 'C_course-v1:HIT+13SC20301820+2019_T1': 'T_孙洁',
 'C_course-v1:TsinghuaX+40050455_2X+sp': 'T_黄霞',
 'C_course-v1:nxu+2018122711+2018_T2': 'T_段玉泉',
 'C_course-v1:TsinghuaX+60610231+2016_T2_SP': 'T_王  前',
 'C_course-v1:Tsin

In [ ]:
dict_lookup_course_teacher['0'] = '0'

In [ ]:
len(dict_lookup_course_teacher)

652

In [ ]:
list(set(list(teacher_course_relation['course'])).symmetric_difference(set(course_sample)))

['C_course-v1:HBNU+2019051509X+2019_T1',
 'C_course-v1:TsinghuaX+AP000003X+2019_T1',
 'C_course-v1:TsinghuaX+70167012X+sp',
 'C_course-v1:SYJU+030020406+2019_T1',
 'C_course-v1:Tsinghua+20150001+sp',
 'C_course-v1:WellesleyX+HIST229x+sp',
 'C_course-v1:JSUX+2017011101X+sp',
 'C_course-v1:Tsinghua+Thesis2017X+2017_T1',
 'C_course-v1:Tsinghua+20181011X+2018_T2']

Check the difference between the courses in the "course and teacher relation" dataframe and course sample. All the course above represent a course without teacher name.

In [ ]:
dict_lookup_course_teacher['C_course-v1:Tsinghua+Thesis2017X+2017_T1'] = 'T_研究生院'
dict_lookup_course_teacher['C_course-v1:TsinghuaX+AP000003X+2019_T1'] = 'T_李强'
dict_lookup_course_teacher['C_course-v1:Tsinghua+20150001+sp'] = 'T_程新兵'
dict_lookup_course_teacher['C_course-v1:Tsinghua+20181011X+2018_T2'] = 'T_谢德军'
dict_lookup_course_teacher['C_course-v1:SYJU+030020406+2019_T1'] = 'T_姜桂荣'
dict_lookup_course_teacher['C_course-v1:TsinghuaX+70167012X+sp'] = 'T_朱颖心'
dict_lookup_course_teacher['C_course-v1:WellesleyX+HIST229x+sp'] = 'T_Adam Van Arsdale'
dict_lookup_course_teacher['C_course-v1:JSUX+2017011101X+sp'] = 'T_李芳宇'
dict_lookup_course_teacher['C_course-v1:HBNU+2019051509X+2019_T1'] = 'T_于丹'

Add a teacher name for the course without the teacher name using a similar teacher from the same school.

The dictionary lookup is constructed.

In [ ]:
for i in course_columns:
  data_teacher[i] = [dict_lookup_course_teacher[item] for item in data_teacher[i]]

Encode the course name with the school name.

In [ ]:
teacher_columns = ['teacher_1', 'teacher_2', 'teacher_3', 'teacher_4', 'teacher_5', 'teacher_6', 'teacher_7', 'teacher_8', 'teacher_9', 'teacher_10']

Named the columns

In [ ]:
data_teacher.columns = teacher_columns

In [ ]:
data_teacher

teacher_1       teacher_2 teacher_3     teacher_4 teacher_5  \
0            0            T_郑莉     T_扈志明         T_陈为蓬      T_向勇   
1            0               0         0             0      T_安宇   
2            0               0         0             0         0   
3            0               0     T_扈志明         T_宫崎泉      T_袁博   
4            0               0         0             0         0   
...        ...             ...       ...           ...       ...   
8956         0               0      T_曾鸣         T_任艳频      T_唐仙   
8957         0               0         0             0     T_彭凯平   
8958         0               0         0             0      T_纪平   
8959         0  T_清华大学自动化系控制课组      T_李琳  T_Bill Aulet     T_王振民   
8960         0               0         0             0         0   

             teacher_6        teacher_7      teacher_8           teacher_9  \
0                 T_安宇             T_吕嵘           T_安宇                T_吕嵘   
1                T_高云峰             T_安宇         T_邢广军                 T_郑莉   
2                 T_彭林            T_清花道          T_程新兵  T_Matthew Thompson   
3                 T_马辉  T_Reid Whitaker           T_安宇                T_安宇   
4                 T_吴烨            T_王曙燕          T_邓俊辉     T_Karine Mardon   
...                ...              ...            ...                 ...   
8956             T_吴华强             T_骆斌           T_孙晶               T_谢德军   
8957             T_戴俊彪     T_Bill Aulet  T_John Guttag                T_郑莉   
8958  T_Sander Kersten        T_小程序慕课讲师  T_John Guttag   T_Bernard Freeman   
8959             T_王晓朝            T_谢德军          T_田素华       T_John Guttag   
8960             T_宫崎泉            T_任艳频         T_邢广军                T_华成英   

              teacher_10  
0                   T_郑莉  
1                   T_胥斌  
2                  T_清花道  
3                  T_扈志明  
4                  T_杨俊荣  
...                  ...  
8956               T_王贵祥  
8957  T_Matthew Thompson  
8958                T_龙瀛  
8959              T_研究生院  
8960               T_王振民  

[8961 rows x 10 columns]

The school data consists of 10 columns from the first school until the final school order for each student (number or row represent number of stundent).

## Label Encoding

### Course Label Dictionary

In [ ]:
course_name = [i for i in np.unique(course_sample)]
course_number = [i for i in range(1, len(np.unique((course_sample)))+1)]
course_encode = {course_name[i]: course_number[i] for i in range(len(course_name))}

course_encode["0"] = 0
course_encode

{'C_course-v1:ACCA+FA1_X+2019_T1': 1,
 'C_course-v1:ACCA+FA1_X+sp': 2,
 'C_course-v1:ACCA+FA1_X_en+2019_T1': 3,
 'C_course-v1:ACCA+FA2_X+2019_T1': 4,
 'C_course-v1:ACCA+FA2_X+sp': 5,
 'C_course-v1:ACCA+FA2_X_en+2019_T1': 6,
 'C_course-v1:ACCA+MA1_X+2019_T1': 7,
 'C_course-v1:ACCA+MA1_X+sp': 8,
 'C_course-v1:ACCA+MA1_X_en+2019_T1': 9,
 'C_course-v1:ACCA+MA2_X+2019_T1': 10,
 'C_course-v1:ACCA+MA2_X+sp': 11,
 'C_course-v1:ACCA+MA2_X_en+2019_T1': 12,
 'C_course-v1:AdelaideX+Wine101x+sp': 13,
 'C_course-v1:AdelaideX+humbio101+sp': 14,
 'C_course-v1:BFU+15002360+2019_T1': 15,
 'C_course-v1:BFU+15012510+2019_T1': 16,
 'C_course-v1:BFU+15023710+2019_T1': 17,
 'C_course-v1:BFU+2018122709+2018_T2': 18,
 'C_course-v1:BFU+2018122709+2019_T1': 19,
 'C_course-v1:BFU+2018122710+2019_T1': 20,
 'C_course-v1:BIFT+1301990078+2019_T1': 21,
 'C_course-v1:BIFT+2018122901X+2018_T2': 22,
 'C_course-v1:BIFT+2018122902X+2018_T2': 23,
 'C_course-v1:BIT+100070018+2019_T1': 24,
 'C_course-v1:BIT+100070018+2019_T2'

### School Label Dictionary

In [ ]:
school_name = [i for i in np.unique(list(school_course_relation['school']))]
school_number = [i for i in range(1, len(np.unique(list(school_course_relation['school'])))+1)]
school_encode = {school_name[i]: school_number[i] for i in range(len(school_name))}

school_encode["0"] = 0
school_encode

{'S_ACCA': 1,
 'S_AdelaideX': 2,
 'S_BFU': 3,
 'S_BIFT': 4,
 'S_BIT': 5,
 'S_BJTU': 6,
 'S_BJUT': 7,
 'S_BNU': 8,
 'S_BSU': 9,
 'S_BTBU': 10,
 'S_BUCM': 11,
 'S_BerkeleyX': 12,
 'S_BurgundyX': 13,
 'S_CCI': 14,
 'S_CHD': 15,
 'S_CIE': 16,
 'S_CMU': 17,
 'S_CSU': 18,
 'S_CUC': 19,
 'S_CUNY': 20,
 'S_CUP': 21,
 'S_CUPB': 22,
 'S_CWU': 23,
 'S_CityU': 24,
 'S_DUT': 25,
 'S_DYU': 26,
 'S_DelftX': 27,
 'S_EST': 28,
 'S_FAFU': 29,
 'S_FJNU': 30,
 'S_FJTCM': 31,
 'S_FUDANx': 32,
 'S_FZXY': 33,
 'S_GDUT': 34,
 'S_GIT': 35,
 'S_GXUST': 36,
 'S_GZHU': 37,
 'S_GZLIS': 38,
 'S_GZUCM': 39,
 'S_HAUT': 40,
 'S_HBNU': 41,
 'S_HBPU': 42,
 'S_HEBNU': 43,
 'S_HEBUT': 44,
 'S_HIT': 45,
 'S_HLJUX': 46,
 'S_HNU': 47,
 'S_HQU': 48,
 'S_HRBEU': 49,
 'S_HUBU': 50,
 'S_HUST': 51,
 'S_ICx': 52,
 'S_IMUN': 53,
 'S_JLU': 54,
 'S_JNU': 55,
 'S_JNUX': 56,
 'S_JSUX': 57,
 'S_JXUST': 58,
 'S_JinanX': 59,
 'S_KMUSTX': 60,
 'S_LUIBE': 61,
 'S_LZU': 62,
 'S_LinuxFoundationX': 63,
 'S_MIL': 64,
 'S_MIT_SLT_Lab': 65,
 'S_M

### Teacher Label Dictionary

In [ ]:
teacher_unique = np.unique(list(dict_lookup_course_teacher.values()))
teacher_name = [i for i in teacher_unique if i != "0"]
teacher_number = [i for i in range(1, len(teacher_name)+1)]
teacher_encode = {teacher_name[i]: teacher_number[i] for i in range(len(teacher_name))}

teacher_encode["0"] = 0
teacher_encode

{'T_Adam Van Arsdale': 1,
 'T_Alma Moon Novotny': 2,
 'T_Armando Fox': 3,
 'T_Arno Smets': 4,
 'T_Barry Walsh': 5,
 'T_Bernard Freeman': 6,
 'T_Bill Aulet': 7,
 'T_Dr. Ariel Fenster': 8,
 'T_Gerald Sussman': 9,
 'T_Howard Takiff': 10,
 'T_J. Kim Vandiver': 11,
 'T_James Donald': 12,
 'T_Jason H. Hafner': 13,
 'T_Jayne Howson': 14,
 'T_Jeffrey D. Sachs': 15,
 'T_John Guttag': 16,
 'T_John Stix': 17,
 'T_Josie Adams': 18,
 'T_Karine Mardon': 19,
 'T_Kelvy Bird': 20,
 'T_Kevin Hsu': 21,
 'T_Kirsten Ostherr': 22,
 'T_Laura Zuheros': 23,
 'T_Lawrence Susskind': 24,
 'T_Maggie Sokolik': 25,
 'T_Marielle ADRIAN': 26,
 'T_Mario Ricci': 27,
 'T_Mark Horswill': 28,
 'T_Matthew Thompson': 29,
 'T_Paul Grbin': 30,
 'T_Pete Mumby': 31,
 'T_Ravi Ramamoorthi': 32,
 'T_Reid Whitaker': 33,
 'T_Sander Kersten': 34,
 'T_Tara Askham': 35,
 'T_Valli Rajagopal': 36,
 'T_丁雅琼': 37,
 'T_乔林': 38,
 'T_于丹': 39,
 'T_于汐': 40,
 'T_于泽': 41,
 'T_于玉贞': 42,
 'T_井福荣': 43,
 'T_任园春': 44,
 'T_任梦一': 45,
 'T_任艳频': 46,
 'T_伍晓明

### Labelling The Data

In [ ]:
def encode_decode_data(data_name, column_name, dict_name):
  for i in column_name:
      data_name[i] = [dict_name[item] for item in data_name[i]]

  return data_name

In [ ]:
data_course = encode_decode_data(data_course, course_columns, course_encode)
data_teacher = encode_decode_data(data_teacher, teacher_columns, teacher_encode)
data_school = encode_decode_data(data_school, school_columns, school_encode)

In [ ]:
data_course

course_1  course_2  course_3  course_4  course_5  course_6  course_7  \
0            0       348       366       331       437       370       420   
1            0         0         0         0       370       415       371   
2            0         0         0         0         0       332       326   
3            0         0       364       408       523       368       259   
4            0         0         0         0         0       463       604   
...        ...       ...       ...       ...       ...       ...       ...   
8956         0         0       556       413       535       441       241   
8957         0         0         0         0       449       548       209   
8958         0         0         0         0        95       599       399   
8959         0       504       595       209       507       340       484   
8960         0         0         0         0         0       408       413   

      course_8  course_9  course_10  
0          371       419        351  
1          405       348         78  
2          309       586        327  
3          370       371        366  
4          434       582        275  
...        ...       ...        ...  
8956       344       484        517  
8957       215       348        586  
8958       215       270        496  
8959       119       215        558  
8960       405       411        507  

[8961 rows x 10 columns]

In [ ]:
data_teacher

teacher_1  teacher_2  teacher_3  teacher_4  teacher_5  teacher_6  \
0             0        402        190        419         97        139   
1             0          0          0          0        139        456   
2             0          0          0          0          0        175   
3             0          0        190        143        365        453   
4             0          0          0          0          0        109   
...         ...        ...        ...        ...        ...        ...   
8956          0          0        198         46        117        102   
8957          0          0          0          0        172        188   
8958          0          0          0          0        337         34   
8959          0        274        232          7        296        298   
8960          0          0          0          0          0        143   

      teacher_7  teacher_8  teacher_9  teacher_10  
0            98        139         98         402  
1           139        395        402         352  
2           275        330         29         275  
3            33        139        139         190  
4           299        392         19         245  
...         ...        ...        ...         ...  
8956        455        129        373         310  
8957          7         16        402          29  
8958        144         16          6         476  
8959        373        323         16         328  
8960         46        395         88         296  

[8961 rows x 10 columns]

In [ ]:
data_school

school_1  school_2  school_3  school_4  school_5  school_6  school_7  \
0            0       103       103       103       103       103       103   
1            0         0         0         0       103       103       103   
2            0         0         0         0         0       103       103   
3            0         0       103       103       103       103        86   
4            0         0         0         0         0       103       115   
...        ...       ...       ...       ...       ...       ...       ...   
8956         0         0       103       103       103       103        76   
8957         0         0         0         0       103       103        66   
8958         0         0         0         0        20       112       103   
8959         0       103       111        66       103       103       103   
8960         0         0         0         0         0       103       103   

      school_8  school_9  school_10  
0          103       103        103  
1          103       103        130  
2          101       107        103  
3          103       103        103  
4          103       107         88  
...        ...       ...        ...  
8956       103       103        103  
8957        66       103        107  
8958        66        86        103  
8959        32        66        103  
8960       103       103        103  

[8961 rows x 10 columns]

Encode the data using label encoding technique using dictionary

### Inverse dictonary for decoding

In [ ]:
course_decode = {y: x for x, y in course_encode.items()}
teacher_decode = {y: x for x, y in teacher_encode.items()}
school_decode = {y: x for x, y in school_encode.items()}

### Main data For Modelling:

In [ ]:
data_course

course_1  course_2  course_3  course_4  course_5  course_6  course_7  \
0            0       348       366       331       437       370       420   
1            0         0         0         0       370       415       371   
2            0         0         0         0         0       332       326   
3            0         0       364       408       523       368       259   
4            0         0         0         0         0       463       604   
...        ...       ...       ...       ...       ...       ...       ...   
8956         0         0       556       413       535       441       241   
8957         0         0         0         0       449       548       209   
8958         0         0         0         0        95       599       399   
8959         0       504       595       209       507       340       484   
8960         0         0         0         0         0       408       413   

      course_8  course_9  course_10  
0          371       419        351  
1          405       348         78  
2          309       586        327  
3          370       371        366  
4          434       582        275  
...        ...       ...        ...  
8956       344       484        517  
8957       215       348        586  
8958       215       270        496  
8959       119       215        558  
8960       405       411        507  

[8961 rows x 10 columns]

## Data Splitting

In [ ]:
X = data_course.drop('course_10', axis=1)
Y = data_course['course_10']

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size = 0.2, random_state = 42)

In [ ]:
rows_train = list(X_train.index)
rows_test = list(X_test.index)

# Data Understanding & Preprocessing - CB

## Data Gathering

In [ ]:
course_info_data = pd.read_json('/content/drive/MyDrive/MOOCCube/entities/course.json', lines=True)
course_info_data

id                   name  \
0             C_course-v1:McGillX+ATOC185x+2015_T1             自然灾害（自主模式）   
1        C_course-v1:TsinghuaX+THESIS2015X+2015_T1  2015年清华大学研究生学位论文答辩（二）   
2    C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_  2014年清华大学研究生学位论文答辩（一）   
3     C_course-v1:TsinghuaX+THESIS2015X_tv+2015_T1  2015年清华大学研究生学位论文答辩（一）   
4                C_course-v1:TsinghuaX+00690242+sp        文物精品与文化中国（自主模式）   
..                                             ...                    ...   
701             C_course-v1:TW+2018052501X+2019_T3   STEM课程设计与案例分析（2019秋）   
702        C_course-v1:TsinghuaX+60700052X+2019_T2          数据科学导论（2019夏）   
703        C_course-v1:TsinghuaX+80515182X+2019_sp         麦肯锡“全球领导力”自主模式   
704            C_course-v1:NEU+2019012201X+2019_T2         物理化学（上）（2019夏）   
705            C_course-v1:NEU+2019012202X+2019_T2         物理化学（下）（2019夏）   

                                  prerequisites  \
0                                             无   
1                                         无先修要求   
2                                         无先修要求   
3                                         无先修要求   
4                                             无   
..                                          ...   
701                                           无   
702                             本课程适合各类学科背景学生学习   
703  Basic knowledge of business administration   
704                                        高等数学   
705                                        高等数学   

                                                 about  \
0    <p>地球上没有一处地方不发生自然灾害。当我们以科学的眼光看待这些自然灾害的原因和本质时，我...   
1    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...   
2    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...   
3    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...   
4    <p>中国考古学是以往100年中发展最为迅速的领域之一，大批珍贵文物的出土，不断刷新人们对文...   
..                                                 ...   
701  <p>\r\n\t《STEM课程设计与案例分析》课程，它是：<br />\r\n1.国内领先...   
702  本课程作为数据科学的先导课和认知类课程，致力于以形象生动的教学模式为学生普及数据挖掘、大数据...   
703  <p>\r\n\t脱胎于清华经管学院与麦肯锡公司联合开设的线下金牌课程，麦肯锡全球领导力在线...   
704  物理化学（上）是一门运用物理学手段和技术研究化学问题的学科，被誉为化学学科的“大脑”。本课程...   
705  物理化学（下）是一门运用物理学手段和技术研究化学问题的学科，被誉为化学学科的“大脑”。本课程...   

                                           core_id  \
0             C_course-v1:McGillX+ATOC185x+2015_T1   
1        C_course-v1:TsinghuaX+THESIS2015X+2015_T1   
2    C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_   
3        C_course-v1:TsinghuaX+THESIS2015X+2015_T1   
4                C_course-v1:TsinghuaX+00690242+sp   
..                                             ...   
701             C_course-v1:TW+2018052501X+2019_T3   
702        C_course-v1:TsinghuaX+60700052X+2019_T1   
703        C_course-v1:TsinghuaX+80515182X+2019_sp   
704            C_course-v1:NEU+2019012201X+2019_T2   
705            C_course-v1:NEU+2019012202X+2019_T2   

                                           video_order  \
0    [V_f6f710068b994452885b90e11b6ee5c5, V_7339568...   
1    [V_de0371575a9f4b5391c89ad16d68b5c2, V_d632034...   
2    [V_d530be9cc0584317a16706684577a6dd, V_f329a62...   
3    [V_de0371575a9f4b5391c89ad16d68b5c2, V_78a8b41...   
4    [V_d7dbd0fe8f504e7a91d863cd5a19b185, V_4492eca...   
..                                                 ...   
701  [V_8f28c813854c444f98950e5404301f0b, V_f23a8ad...   
702  [V_2aa0936927744c3b9f1dcc8b5bc30fba, V_2def333...   
703  [V_ab97fe9d9a744958b886aa74d44146f8, V_8973856...   
704  [V_6595d9e0894848348287b8abaa30dce6, V_4dc4a13...   
705  [V_c21bcc55e0b84c539fb874cac8b24b68, V_8f1f9bb...   

                                          display_name  \
0    [Video: Overview 1, Video: Overview 2, Video: ...   
1    [答辩陈述, 答辩陈述, 问答及答辩结果, 答辩陈述, 问答及答辩结果, 答辩陈述, 问答及...   
2    [论文答辩实况, 问答及答辩结果, 导师评价, 同学眼中的王鑫, 个人学术感言, 吴宇恩答辩...   
3    [答辩陈述, 问答及答辩结果, 答辩陈述, 问答及答辩结果, 答辩陈述, 问答及答辩结果, ...   
4    [文献所见原始渡河工具, 《禹贡》所见的水路交通, 绰墩山出土的渡河浮木, 舟船考古发掘, ...   
..                                                 ...   
701  [何为 STEM 教育, STEM 教育的起源与发展, 世界各国的 STEM 教育政策, 为...   
702

In [ ]:
course_info_data = course_info_data[['id','name', 'prerequisites', 'about']]

Rename the columns

In [ ]:
course_info_data

id                   name  \
0             C_course-v1:McGillX+ATOC185x+2015_T1             自然灾害（自主模式）   
1        C_course-v1:TsinghuaX+THESIS2015X+2015_T1  2015年清华大学研究生学位论文答辩（二）   
2    C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_  2014年清华大学研究生学位论文答辩（一）   
3     C_course-v1:TsinghuaX+THESIS2015X_tv+2015_T1  2015年清华大学研究生学位论文答辩（一）   
4                C_course-v1:TsinghuaX+00690242+sp        文物精品与文化中国（自主模式）   
..                                             ...                    ...   
701             C_course-v1:TW+2018052501X+2019_T3   STEM课程设计与案例分析（2019秋）   
702        C_course-v1:TsinghuaX+60700052X+2019_T2          数据科学导论（2019夏）   
703        C_course-v1:TsinghuaX+80515182X+2019_sp         麦肯锡“全球领导力”自主模式   
704            C_course-v1:NEU+2019012201X+2019_T2         物理化学（上）（2019夏）   
705            C_course-v1:NEU+2019012202X+2019_T2         物理化学（下）（2019夏）   

                                  prerequisites  \
0                                             无   
1                                         无先修要求   
2                                         无先修要求   
3                                         无先修要求   
4                                             无   
..                                          ...   
701                                           无   
702                             本课程适合各类学科背景学生学习   
703  Basic knowledge of business administration   
704                                        高等数学   
705                                        高等数学   

                                                 about  
0    <p>地球上没有一处地方不发生自然灾害。当我们以科学的眼光看待这些自然灾害的原因和本质时，我...  
1    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...  
2    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...  
3    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...  
4    <p>中国考古学是以往100年中发展最为迅速的领域之一，大批珍贵文物的出土，不断刷新人们对文...  
..                                                 ...  
701  <p>\r\n\t《STEM课程设计与案例分析》课程，它是：<br />\r\n1.国内领先...  
702  本课程作为数据科学的先导课和认知类课程，致力于以形象生动的教学模式为学生普及数据挖掘、大数据...  
703  <p>\r\n\t脱胎于清华经管学院与麦肯锡公司联合开设的线下金牌课程，麦肯锡全球领导力在线...  
704  物理化学（上）是一门运用物理学手段和技术研究化学问题的学科，被誉为化学学科的“大脑”。本课程...  
705  物理化学（下）是一门运用物理学手段和技术研究化学问题的学科，被誉为化学学科的“大脑”。本课程...  

[706 rows x 4 columns]

## Features Construction

In [ ]:
add_column = list(course_info_data['id'])
add_column

['C_course-v1:McGillX+ATOC185x+2015_T1',
 'C_course-v1:TsinghuaX+THESIS2015X+2015_T1',
 'C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_',
 'C_course-v1:TsinghuaX+THESIS2015X_tv+2015_T1',
 'C_course-v1:TsinghuaX+00690242+sp',
 'C_course-v1:HNU+HNU001+sp',
 'C_course-v1:TsinghuaX+00720091X+sp',
 'C_course-v1:TsinghuaX+00612642X+sp',
 'C_course-v1:TsinghuaX+00691153X+sp',
 'C_course-v1:UQx+Think101x+sp',
 'C_course-v1:TsinghuaX+JWWCD001+2015_T2',
 'C_course-v1:TsinghuaX+70660542X+2015_T2',
 'C_course-v1:TsinghuaX+70120073X+sp',
 'C_course-v1:PekingX+Peking001x+_',
 'C_course-v1:UST+UST001+sp',
 'C_course-v1:SDUx+00931800X+sp',
 'C_course-v1:TsinghuaX+00690212X+sp',
 'C_course-v1:TsinghuaX+00680082X+sp',
 'C_course-v1:TsinghuaX+30260112X+sp',
 'C_course-v1:TsinghuaX+00310222X+sp',
 'C_course-v1:BIT+PHY1701702+sp',
 'C_course-v1:BIT+PHY1701701+sp',
 'C_course-v1:TsinghuaX+80515522X+sp',
 'C_course-v1:TsinghuaX+34000888X+sp',
 'C_course-v1:TsinghuaX+00690092X+sp',
 'C_course-v1:TsinghuaX+0069

In [ ]:
course_info_data['teacher'] = add_column
course_info_data['school'] = add_column

<ipython-input-213-33bda7fde19b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  course_info_data['teacher'] = add_column
<ipython-input-213-33bda7fde19b>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  course_info_data['school'] = add_column


In [ ]:
course_info_data

id                   name  \
0             C_course-v1:McGillX+ATOC185x+2015_T1             自然灾害（自主模式）   
1        C_course-v1:TsinghuaX+THESIS2015X+2015_T1  2015年清华大学研究生学位论文答辩（二）   
2    C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_  2014年清华大学研究生学位论文答辩（一）   
3     C_course-v1:TsinghuaX+THESIS2015X_tv+2015_T1  2015年清华大学研究生学位论文答辩（一）   
4                C_course-v1:TsinghuaX+00690242+sp        文物精品与文化中国（自主模式）   
..                                             ...                    ...   
701             C_course-v1:TW+2018052501X+2019_T3   STEM课程设计与案例分析（2019秋）   
702        C_course-v1:TsinghuaX+60700052X+2019_T2          数据科学导论（2019夏）   
703        C_course-v1:TsinghuaX+80515182X+2019_sp         麦肯锡“全球领导力”自主模式   
704            C_course-v1:NEU+2019012201X+2019_T2         物理化学（上）（2019夏）   
705            C_course-v1:NEU+2019012202X+2019_T2         物理化学（下）（2019夏）   

                                  prerequisites  \
0                                             无   
1                                         无先修要求   
2                                         无先修要求   
3                                         无先修要求   
4                                             无   
..                                          ...   
701                                           无   
702                             本课程适合各类学科背景学生学习   
703  Basic knowledge of business administration   
704                                        高等数学   
705                                        高等数学   

                                                 about  \
0    <p>地球上没有一处地方不发生自然灾害。当我们以科学的眼光看待这些自然灾害的原因和本质时，我...   
1    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...   
2    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...   
3    <p>学位论文答辩环节是研究生培养的重要环节，为了充分发挥该环节的育人作用，搭建学术交流的平...   
4    <p>中国考古学是以往100年中发展最为迅速的领域之一，大批珍贵文物的出土，不断刷新人们对文...   
..                                                 ...   
701  <p>\r\n\t《STEM课程设计与案例分析》课程，它是：<br />\r\n1.国内领先...   
702  本课程作为数据科学的先导课和认知类课程，致力于以形象生动的教学模式为学生普及数据挖掘、大数据...   
703  <p>\r\n\t脱胎于清华经管学院与麦肯锡公司联合开设的线下金牌课程，麦肯锡全球领导力在线...   
704  物理化学（上）是一门运用物理学手段和技术研究化学问题的学科，被誉为化学学科的“大脑”。本课程...   
705  物理化学（下）是一门运用物理学手段和技术研究化学问题的学科，被誉为化学学科的“大脑”。本课程...   

                                           teacher  \
0             C_course-v1:McGillX+ATOC185x+2015_T1   
1        C_course-v1:TsinghuaX+THESIS2015X+2015_T1   
2    C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_   
3     C_course-v1:TsinghuaX+THESIS2015X_tv+2015_T1   
4                C_course-v1:TsinghuaX+00690242+sp   
..                                             ...   
701             C_course-v1:TW+2018052501X+2019_T3   
702        C_course-v1:TsinghuaX+60700052X+2019_T2   
703        C_course-v1:TsinghuaX+80515182X+2019_sp   
704            C_course-v1:NEU+2019012201X+2019_T2   
705            C_course-v1:NEU+2019012202X+2019_T2   

                                            school  
0             C_course-v1:McGillX+ATOC185x+2015_T1  
1        C_course-v1:TsinghuaX+THESIS2015X+2015_T1  
2    C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_  
3     C_course-v1:TsinghuaX+THESIS2015X_tv+2015_T1  
4                C_course-v1:TsinghuaX+00690242+sp  
..                                             ...  
701             C_course-v1:TW+2018052501X+2019_T3  
702        C_course-v1:TsinghuaX+60700052X+2019_T2  
703        C_course-v1:TsinghuaX+80515182X+2019_sp  
704            C_course-v1:NEU+2019012201X+2019_T2  
705            C_course-v1:NEU+2019012202X+2019_T2  

[706 rows x 6 columns]

Add columns teacher and school. Right now the teacher and school columns consist of course ID, later the course ID will be change to teacher ID for teacher column and school ID for school column by using respective dictionary.

In [ ]:
teacher_course_relation = pd.read_csv('/content/drive/MyDrive/MOOCCube/relations/teacher-course.json', sep='\t', header=None)
teacher_course_relation

0                                          1
0     T_方维奇            C_course-v1:SPI+20170828001x+sp
1     T_方维奇      C_course-v1:SXPI+20170828001x+2019_T1
2     T_范茂魁      C_course-v1:PSFFC+2018102405X+2018_T2
3     T_连小珉  C_course-v1:TsinghuaX+70150104_2X+2019_T1
4     T_连小珉    C_course-v1:TsinghuaX+70150104X+2019_T1
...     ...                                        ...
2344  T_徐君莉        C_course-v1:NEU+2018122401X+2019_T1
2345   T_薛庆          C_course-v1:BIT+100070018+2019_T2
2346  T_史静琤          C_course-v1:CSU+20180919X+2019_T1
2347  T_刘小冰         C_course-v1:TsinghuaX+80515522X+sp
2348  T_黄湘晖        C_course-v1:XMU+2019031401X+2019_T1

[2349 rows x 2 columns]

In [ ]:
teacher_course_relation.set_axis(['teacher', 'course'], axis='columns', inplace=True)

<ipython-input-217-44f424829db4>:1: FutureWarning: DataFrame.set_axis 'inplace' keyword is deprecated and will be removed in a future version. Use `obj = obj.set_axis(..., copy=False)` instead
  teacher_course_relation.set_axis(['teacher', 'course'], axis='columns', inplace=True)


In [ ]:
teacher_course_relation

teacher                                     course
0      T_方维奇            C_course-v1:SPI+20170828001x+sp
1      T_方维奇      C_course-v1:SXPI+20170828001x+2019_T1
2      T_范茂魁      C_course-v1:PSFFC+2018102405X+2018_T2
3      T_连小珉  C_course-v1:TsinghuaX+70150104_2X+2019_T1
4      T_连小珉    C_course-v1:TsinghuaX+70150104X+2019_T1
...      ...                                        ...
2344   T_徐君莉        C_course-v1:NEU+2018122401X+2019_T1
2345    T_薛庆          C_course-v1:BIT+100070018+2019_T2
2346   T_史静琤          C_course-v1:CSU+20180919X+2019_T1
2347   T_刘小冰         C_course-v1:TsinghuaX+80515522X+sp
2348   T_黄湘晖        C_course-v1:XMU+2019031401X+2019_T1

[2349 rows x 2 columns]

In [ ]:
teacher_course_relation = teacher_course_relation.iloc[:,[1, 0]]

In [ ]:
teacher_course_relation

course teacher
0               C_course-v1:SPI+20170828001x+sp   T_方维奇
1         C_course-v1:SXPI+20170828001x+2019_T1   T_方维奇
2         C_course-v1:PSFFC+2018102405X+2018_T2   T_范茂魁
3     C_course-v1:TsinghuaX+70150104_2X+2019_T1   T_连小珉
4       C_course-v1:TsinghuaX+70150104X+2019_T1   T_连小珉
...                                         ...     ...
2344        C_course-v1:NEU+2018122401X+2019_T1   T_徐君莉
2345          C_course-v1:BIT+100070018+2019_T2    T_薛庆
2346          C_course-v1:CSU+20180919X+2019_T1   T_史静琤
2347         C_course-v1:TsinghuaX+80515522X+sp   T_刘小冰
2348        C_course-v1:XMU+2019031401X+2019_T1   T_黄湘晖

[2349 rows x 2 columns]

In [ ]:
len(np.unique(teacher_course_relation['course']))

697

In [ ]:
school_course_relation = pd.read_csv('/content/drive/MyDrive/MOOCCube/relations/school-course.json', sep='\t', header=None)
school_course_relation

0                                      1
0      S_BNU    C_course-v1:BNU+CSL21148501+2018_T2
1      S_BNU    C_course-v1:BNU+GE310141091+2019_T1
2      S_BNU    C_course-v1:BNU+2017112001X+2019_T1
3      S_BNU    C_course-v1:BNU+2018122603X+2018_T2
4      S_BNU    C_course-v1:BNU+CSL21126882+2019_T1
..       ...                                    ...
700  S_PSFFC  C_course-v1:PSFFC+2018102405X+2018_T2
701  S_PSFFC  C_course-v1:PSFFC+2018102403X+2018_T2
702  S_PSFFC  C_course-v1:PSFFC+2018102406X+2018_T2
703  S_PSFFC  C_course-v1:PSFFC+2018102404X+2018_T2
704    S_QLU    C_course-v1:QLU+2018122607X+2019_T1

[705 rows x 2 columns]

In [ ]:
school_course_relation.set_axis(['school', 'course'], axis='columns', inplace=True)

<ipython-input-223-ce3aad33b220>:1: FutureWarning: DataFrame.set_axis 'inplace' keyword is deprecated and will be removed in a future version. Use `obj = obj.set_axis(..., copy=False)` instead
  school_course_relation.set_axis(['school', 'course'], axis='columns', inplace=True)


In [ ]:
school_course_relation

school                                 course
0      S_BNU    C_course-v1:BNU+CSL21148501+2018_T2
1      S_BNU    C_course-v1:BNU+GE310141091+2019_T1
2      S_BNU    C_course-v1:BNU+2017112001X+2019_T1
3      S_BNU    C_course-v1:BNU+2018122603X+2018_T2
4      S_BNU    C_course-v1:BNU+CSL21126882+2019_T1
..       ...                                    ...
700  S_PSFFC  C_course-v1:PSFFC+2018102405X+2018_T2
701  S_PSFFC  C_course-v1:PSFFC+2018102403X+2018_T2
702  S_PSFFC  C_course-v1:PSFFC+2018102406X+2018_T2
703  S_PSFFC  C_course-v1:PSFFC+2018102404X+2018_T2
704    S_QLU    C_course-v1:QLU+2018122607X+2019_T1

[705 rows x 2 columns]

In [ ]:
school_course_relation = school_course_relation.iloc[:,[1, 0]]

In [ ]:
school_course_relation

course   school
0      C_course-v1:BNU+CSL21148501+2018_T2    S_BNU
1      C_course-v1:BNU+GE310141091+2019_T1    S_BNU
2      C_course-v1:BNU+2017112001X+2019_T1    S_BNU
3      C_course-v1:BNU+2018122603X+2018_T2    S_BNU
4      C_course-v1:BNU+CSL21126882+2019_T1    S_BNU
..                                     ...      ...
700  C_course-v1:PSFFC+2018102405X+2018_T2  S_PSFFC
701  C_course-v1:PSFFC+2018102403X+2018_T2  S_PSFFC
702  C_course-v1:PSFFC+2018102406X+2018_T2  S_PSFFC
703  C_course-v1:PSFFC+2018102404X+2018_T2  S_PSFFC
704    C_course-v1:QLU+2018122607X+2019_T1    S_QLU

[705 rows x 2 columns]

In [ ]:
school_course_relation

course   school
0      C_course-v1:BNU+CSL21148501+2018_T2    S_BNU
1      C_course-v1:BNU+GE310141091+2019_T1    S_BNU
2      C_course-v1:BNU+2017112001X+2019_T1    S_BNU
3      C_course-v1:BNU+2018122603X+2018_T2    S_BNU
4      C_course-v1:BNU+CSL21126882+2019_T1    S_BNU
..                                     ...      ...
700  C_course-v1:PSFFC+2018102405X+2018_T2  S_PSFFC
701  C_course-v1:PSFFC+2018102403X+2018_T2  S_PSFFC
702  C_course-v1:PSFFC+2018102406X+2018_T2  S_PSFFC
703  C_course-v1:PSFFC+2018102404X+2018_T2  S_PSFFC
704    C_course-v1:QLU+2018122607X+2019_T1    S_QLU

[705 rows x 2 columns]

In [ ]:
dict_lookup_course_school = dict(zip(school_course_relation['course'], school_course_relation['school']))
dict_lookup_course_school

{'C_course-v1:BNU+CSL21148501+2018_T2': 'S_BNU',
 'C_course-v1:BNU+GE310141091+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2017112001X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018122603X+2018_T2': 'S_BNU',
 'C_course-v1:BNU+CSL21126882+2019_T1': 'S_BNU',
 'C_course-v1:BNU+0610073991+2019_T1': 'S_BNU',
 'C_course-v1:BNU+1010070372+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2017053101X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+GE410081071-01+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018122405X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018122602X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091301X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+CHE21128011+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091302X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091303X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+PHI2107404101+2019_T1': 'S_BNU',
 'C_course-v1:BNU+0610073981+2019_T1': 'S_BNU',
 'C_course-v1:BNU+0210021441+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2017071001X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091305X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2

In [ ]:
dict_lookup_course_school['0'] = '0'

In [ ]:
len(dict_lookup_course_school)

706

In [ ]:
dict_lookup_course_school['C_course-v1:TsinghuaX+AP000003X+2019_T1'] = 'S_TsinghuaX'

In [ ]:
dict_lookup_course_teacher = dict(zip(teacher_course_relation['course'], teacher_course_relation['teacher']))
dict_lookup_course_teacher

{'C_course-v1:SPI+20170828001x+sp': 'T_姚常青',
 'C_course-v1:SXPI+20170828001x+2019_T1': 'T_姚常青',
 'C_course-v1:PSFFC+2018102405X+2018_T2': 'T_赵石楠',
 'C_course-v1:TsinghuaX+70150104_2X+2019_T1': 'T_连小珉',
 'C_course-v1:TsinghuaX+70150104X+2019_T1': 'T_连小珉',
 'C_course-v1:TJU+2010241X+sp': 'T_张德顺',
 'C_course-v1:TsinghuaX+20250103X+sp': 'T_任艳频',
 'C_course-v1:KMUSTX+8219011+2019_T1': 'T_王裕森',
 'C_course-v1:BIFT+2018122901X+2018_T2': 'T_丁雅琼',
 'C_course-v1:PSFFC+2018102403X+2018_T2': 'T_陶昆',
 'C_course-v1:BUCM+2018122604X+2018_T2': 'T_周芬',
 'C_course-v1:BUCM+2018122604X+2019_T1': 'T_周芬',
 'C_course-v1:TsinghuaX+20180919X+2019_T1': 'T_靳卫萍',
 'C_course-v1:SCUT+145033+sp': 'T_杨俊荣',
 'C_course-v1:BNU+2018091303X+2019_T1': 'T_孙璞',
 'C_course-v1:BNU+2018091304X+2019_T1': 'T_杨兆春',
 'C_course-v1:HIT+13SC20301820+2019_T1': 'T_孙洁',
 'C_course-v1:TsinghuaX+40050455_2X+sp': 'T_黄霞',
 'C_course-v1:nxu+2018122711+2018_T2': 'T_段玉泉',
 'C_course-v1:nxu+2018122711+2019_T1': 'T_段玉泉',
 'C_course-v1:XJTU+2018122

In [ ]:
dict_lookup_course_teacher['0'] = '0'

In [ ]:
len(dict_lookup_course_teacher)

698

In [ ]:
dict_lookup_course_teacher['C_course-v1:Tsinghua+Thesis2017X+2017_T1'] = 'T_研究生院'
dict_lookup_course_teacher['C_course-v1:TsinghuaX+AP000003X+2019_T1'] = 'T_李强'
dict_lookup_course_teacher['C_course-v1:Tsinghua+20150001+sp'] = 'T_程新兵'
dict_lookup_course_teacher['C_course-v1:Tsinghua+20181011X+2018_T2'] = 'T_谢德军'
dict_lookup_course_teacher['C_course-v1:SYJU+030020406+2019_T1'] = 'T_姜桂荣'
dict_lookup_course_teacher['C_course-v1:TsinghuaX+70167012X+sp'] = 'T_朱颖心'
dict_lookup_course_teacher['C_course-v1:WellesleyX+HIST229x+sp'] = 'T_Adam Van Arsdale'
dict_lookup_course_teacher['C_course-v1:JSUX+2017011101X+sp'] = 'T_李芳宇'
dict_lookup_course_teacher['C_course-v1:HBNU+2019051509X+2019_T1'] = 'T_于丹'

In [ ]:
dict_lookup_course_school

{'C_course-v1:BNU+CSL21148501+2018_T2': 'S_BNU',
 'C_course-v1:BNU+GE310141091+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2017112001X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018122603X+2018_T2': 'S_BNU',
 'C_course-v1:BNU+CSL21126882+2019_T1': 'S_BNU',
 'C_course-v1:BNU+0610073991+2019_T1': 'S_BNU',
 'C_course-v1:BNU+1010070372+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2017053101X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+GE410081071-01+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018122405X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018122602X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091301X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+CHE21128011+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091302X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091303X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+PHI2107404101+2019_T1': 'S_BNU',
 'C_course-v1:BNU+0610073981+2019_T1': 'S_BNU',
 'C_course-v1:BNU+0210021441+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2017071001X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2018091305X+2019_T1': 'S_BNU',
 'C_course-v1:BNU+2

In [ ]:
course_info_data['teacher'] = [dict_lookup_course_teacher[item] for item in course_info_data['teacher']]
course_info_data['school'] = [dict_lookup_course_school[item] for item in course_info_data['school']]

<ipython-input-237-ebad6b21711c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  course_info_data['teacher'] = [dict_lookup_course_teacher[item] for item in course_info_data['teacher']]
<ipython-input-237-ebad6b21711c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  course_info_data['school'] = [dict_lookup_course_school[item] for item in course_info_data['school']]


Transform the course ID in teacher column to teacher ID, transfrom the course ID in school column to school ID.

In [ ]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.cluster import KMeans
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.decomposition import PCA
import re
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import linear_kernel

## Clean Text Data

In [ ]:
course_info_data['name'] = [re.sub(r'[^\w\s]', '', t) for t in course_info_data['name']]
course_info_data['prerequisites'] = [re.sub(',',' ',re.sub(' ','',t)) for t in course_info_data['prerequisites']]
course_info_data['about'] = [re.sub(r'[^\w\s]', '', t) for t in course_info_data['about']]
course_info_data['teacher'] = [re.sub(r'[^\w\s]', '', t) for t in course_info_data['teacher']]
course_info_data['school'] = [re.sub(r'[^\w\s]', '', t) for t in course_info_data['school']]

<ipython-input-239-7fdff98c4a61>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  course_info_data['name'] = [re.sub(r'[^\w\s]', '', t) for t in course_info_data['name']]
<ipython-input-239-7fdff98c4a61>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  course_info_data['prerequisites'] = [re.sub(',',' ',re.sub(' ','',t)) for t in course_info_data['prerequisites']]
<ipython-input-239-7fdff98c4a61>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[ro

In [ ]:
course_info_data.head()

id                 name  \
0           C_course-v1:McGillX+ATOC185x+2015_T1             自然灾害自主模式   
1      C_course-v1:TsinghuaX+THESIS2015X+2015_T1  2015年清华大学研究生学位论文答辩二   
2  C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_  2014年清华大学研究生学位论文答辩一   
3   C_course-v1:TsinghuaX+THESIS2015X_tv+2015_T1  2015年清华大学研究生学位论文答辩一   
4              C_course-v1:TsinghuaX+00690242+sp        文物精品与文化中国自主模式   

  prerequisites                                              about  \
0             无  p地球上没有一处地方不发生自然灾害当我们以科学的眼光看待这些自然灾害的原因和本质时我们可以更...   
1         无先修要求  p学位论文答辩环节是研究生培养的重要环节为了充分发挥该环节的育人作用搭建学术交流的平台进一步...   
2         无先修要求  p学位论文答辩环节是研究生培养的重要环节为了充分发挥该环节的育人作用搭建学术交流的平台进一步...   
3         无先修要求  p学位论文答辩环节是研究生培养的重要环节为了充分发挥该环节的育人作用搭建学术交流的平台进一步...   
4             无  p中国考古学是以往100年中发展最为迅速的领域之一大批珍贵文物的出土不断刷新人们对文化中国的...   

       teacher       school  
0  T_John Stix    S_McGillX  
1       T_研究生院  S_TsinghuaX  
2       T_研究生院  S_TsinghuaX  
3       T_研究生院  S_TsinghuaX  
4         T_彭林  S_TsinghuaX

## Features Integration

In [ ]:
course_info_data["combined"] = course_info_data['name'] + '  ' + course_info_data['prerequisites'] + ' ' + course_info_data['about'] + ' ' + course_info_data['teacher'] + ' ' + course_info_data['school']


<ipython-input-241-16de9ef1c52f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  course_info_data["combined"] = course_info_data['name'] + '  ' + course_info_data['prerequisites'] + ' ' + course_info_data['about'] + ' ' + course_info_data['teacher'] + ' ' + course_info_data['school']


In [ ]:
course_info_data.head()

id                 name  \
0           C_course-v1:McGillX+ATOC185x+2015_T1             自然灾害自主模式   
1      C_course-v1:TsinghuaX+THESIS2015X+2015_T1  2015年清华大学研究生学位论文答辩二   
2  C_course-v1:TsinghuaX+THESIS2014_1X_tv+_2014_  2014年清华大学研究生学位论文答辩一   
3   C_course-v1:TsinghuaX+THESIS2015X_tv+2015_T1  2015年清华大学研究生学位论文答辩一   
4              C_course-v1:TsinghuaX+00690242+sp        文物精品与文化中国自主模式   

  prerequisites                                              about  \
0             无  p地球上没有一处地方不发生自然灾害当我们以科学的眼光看待这些自然灾害的原因和本质时我们可以更...   
1         无先修要求  p学位论文答辩环节是研究生培养的重要环节为了充分发挥该环节的育人作用搭建学术交流的平台进一步...   
2         无先修要求  p学位论文答辩环节是研究生培养的重要环节为了充分发挥该环节的育人作用搭建学术交流的平台进一步...   
3         无先修要求  p学位论文答辩环节是研究生培养的重要环节为了充分发挥该环节的育人作用搭建学术交流的平台进一步...   
4             无  p中国考古学是以往100年中发展最为迅速的领域之一大批珍贵文物的出土不断刷新人们对文化中国的...   

       teacher       school                                           combined  
0  T_John Stix    S_McGillX  自然灾害自主模式  无 p地球上没有一处地方不发生自然灾害当我们以科学的眼光看待这些自然灾害...  
1       T_研究生院  S_TsinghuaX  2015年清华大学研究生学位论文答辩二  无先修要求 p学位论文答辩环节是研究生培养的重要环...  
2       T_研究生院  S_TsinghuaX  2014年清华大学研究生学位论文答辩一  无先修要求 p学位论文答辩环节是研究生培养的重要环...  
3       T_研究生院  S_TsinghuaX  2015年清华大学研究生学位论文答辩一  无先修要求 p学位论文答辩环节是研究生培养的重要环...  
4         T_彭林  S_TsinghuaX  文物精品与文化中国自主模式  无 p中国考古学是以往100年中发展最为迅速的领域之一大批珍贵...

Integrate features of course name, prerequisites, about, teacher and school into one named as "combined".

## Add More Courses Into Dictionary

In [ ]:
course_encode['C_course-v1:nxu+2018122713+2019_T1'] = 661
course_encode['C_course-v1:SDSNAssociation+ECD001+sp'] = 662
course_encode['C_course-v1:SWPU+3615001035+2019_T1'] = 663
course_encode['C_course-v1:LUIBE+201808064+2019_T1'] = 664
course_encode['C_course-v1:nxu+2018122711+2019_T1'] = 665
course_encode['C_course-v1:CIE+JD_2017+2019_T1'] = 666
course_encode['C_course-v1:dlmu+20180906+2019_T1'] = 667
course_encode['C_course-v1:BNU+2018122405X+2019_T1'] = 668
course_encode['C_course-v1:CIE+CIE2016005+2019_T1'] = 669
course_encode['C_course-v1:HIT+HIT2016001+2019_T1'] = 670
course_encode['C_course-v1:SDSNAssociation+PB+sp'] = 671
course_encode['C_course-v1:CIE+CIE2017008+2019_T1'] = 672
course_encode['C_course-v1:FZXY+20180301001+2019_T1'] = 673
course_encode['C_course-v1:CIE+CIE2017005+2019_T1'] = 674
course_encode['C_course-v1:TsinghuaX+80511503X+2019_T1'] = 675
course_encode['C_course-v1:CIE+CIE2017001+2019_T1'] = 676
course_encode['C_course-v1:SDSNAssociation+SC001+sp'] = 677
course_encode['C_course-v1:TJUFE+2018122506X+2019_T1'] = 678
course_encode['C_course-v1:TsinghuaX+01510192X+2019_T1'] = 679
course_encode['C_course-v1:XYSFXY+20181024X+2019_T1'] = 680
course_encode['C_course-v1:HNU+20180424001+2018_T1'] = 681
course_encode['C_course-v1:CSMZXY+2018111301X+2019_T2'] = 682
course_encode['C_course-v1:TsinghuaX+80240372X+sp'] = 683
course_encode['C_course-v1:JNU+11020009+2019_T1'] = 684
course_encode['C_course-v1:BSU+2018122405X+2019_T1'] = 685
course_encode['C_course-v1:SEU+00690803_2+2019_T1'] = 686
course_encode['C_course-v1:PSFFC+2018102404X+2018_T2'] = 687
course_encode['C_course-v1:NEU+2018051501+sp'] = 688
course_encode['C_course-v1:CSU+2019043001X+2019_T1'] = 689
course_encode['C_course-v1:SDSNAssociation+TOW001+sp'] = 690
course_encode['C_course-v1:CIE+CIE2016004+2019_T1'] = 691
course_encode['C_course-v1:SXPI+20171101002+sp'] = 692
course_encode['C_course-v1:KMUSTX+1803168+2018_T2'] = 693
course_encode['C_course-v1:SDUx+00931800X+sp'] = 694
course_encode['C_course-v1:BSU+2018122404X+2019_T1'] = 695
course_encode['C_course-v1:BNU+CHE21128011+2019_T1'] = 696
course_encode['C_course-v1:ZZU+20180116001+2019_T1'] = 697
course_encode['C_course-v1:SDSNAssociation+CSN+sp'] = 698
course_encode['C_course-v1:TsinghuaX+70340063X+2019_T1'] = 699
course_encode['C_course-v1:HEBUT+2019040202X+2019_T1'] = 700
course_encode['C_course-v1:FAFU+55071003+2019_T1'] = 701
course_encode['C_course-v1:XJTU+2018122507X+2018_T2'] = 702
course_encode['C_course-v1:JNU+07009215+2019_T1'] = 703
course_encode['C_course-v1:CIE+CIE2016002+2019_T1'] = 704
course_encode['C_course-v1:BNU+2017112001X+2019_T1'] = 705
course_encode['C_course-v1:HUBU+HU08001X+2019_T1'] = 706

In [ ]:
dict_lookup_course_prerequisite['C_course-v1:SEU+00690803_2+2019_T1'] = ["0"]
dict_lookup_course_prerequisite['C_course-v1:NEU+2018051501+sp']=["0"]
dict_lookup_course_prerequisite['C_course-v1:CIE+CIE2017001+2019_T1'] = ["0"]
dict_lookup_course_prerequisite['C_course-v1:LUIBE+201808064+2019_T1']=["0"]
dict_lookup_course_prerequisite['C_course-v1:SDSNAssociation+SC001+sp'] = ["0"]
dict_lookup_course_prerequisite['C_course-v1:SDSNAssociation+CSN+sp']=["0"]
dict_lookup_course_prerequisite['C_course-v1:SDSNAssociation+ECD001+sp'] = ["0"]
dict_lookup_course_prerequisite['C_course-v1:SDSNAssociation+PB+sp']=["0"]
dict_lookup_course_prerequisite['C_course-v1:BNU+2017112001X+2019_T1'] = ["0"]
dict_lookup_course_prerequisite['C_course-v1:XYSFXY+20181024X+2019_T1']=["0"]
dict_lookup_course_prerequisite['C_course-v1:TsinghuaX+80240372X+sp'] = ["0"]
dict_lookup_course_prerequisite['C_course-v1:XJTU+2018122507X+2018_T2'] = ["0"]
dict_lookup_course_prerequisite['C_course-v1:JNU+11020009+2019_T1'] = ["0"]

dict_lookup_course_prerequisite['C_course-v1:PSFFC+2018102404X+2018_T2']=["0"]
dict_lookup_course_prerequisite['C_course-v1:HUBU+HU08001X+2019_T1'] = ["0"]
dict_lookup_course_prerequisite['C_course-v1:SWPU+3615001035+2019_T1']=["0"]
dict_lookup_course_prerequisite['C_course-v1:TsinghuaX+80511503X+2019_T1'] = ["0"]
dict_lookup_course_prerequisite['C_course-v1:XJTU+2018122507X+2018_T2'] = ["0"]
dict_lookup_course_prerequisite['C_course-v1:JNU+07009215+2019_T1'] = ["0"]

### Inverse dictonary for decoding

In [ ]:
course_decode = {y: x for x, y in course_encode.items()}
teacher_decode = {y: x for x, y in teacher_encode.items()}
school_decode = {y: x for x, y in school_encode.items()}

## Label Encoding

In [ ]:
course_info_data['id'] = [course_encode[item] for item in course_info_data['id']]

<ipython-input-244-ced47b950ede>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  course_info_data['id'] = [course_encode[item] for item in course_info_data['id']]


Encode the course id to integer using dictionary

In [ ]:
course_info_data

id                 name                           prerequisites  \
0    219             自然灾害自主模式                                       无   
1    558  2015年清华大学研究生学位论文答辩二                                   无先修要求   
2    557  2014年清华大学研究生学位论文答辩一                                   无先修要求   
3    559  2015年清华大学研究生学位论文答辩一                                   无先修要求   
4    334        文物精品与文化中国自主模式                                       无   
..   ...                  ...                                     ...   
701  308   STEM课程设计与案例分析2019秋                                       无   
702  492          数据科学导论2019夏                         本课程适合各类学科背景学生学习   
703  525         麦肯锡全球领导力自主模式  Basicknowledgeofbusinessadministration   
704  234           物理化学上2019夏                                    高等数学   
705  235           物理化学下2019夏                                    高等数学   

                                                 about      teacher  \
0    p地球上没有一处地方不发生自然灾害当我们以科学的眼光看待这些自然灾害的原因和本质时我们可以更...  T_John Stix   
1    p学位论文答辩环节是研究生培养的重要环节为了充分发挥该环节的育人作用搭建学术交流的平台进一步...       T_研究生院   
2    p学位论文答辩环节是研究生培养的重要环节为了充分发挥该环节的育人作用搭建学术交流的平台进一步...       T_研究生院   
3    p学位论文答辩环节是研究生培养的重要环节为了充分发挥该环节的育人作用搭建学术交流的平台进一步...       T_研究生院   
4    p中国考古学是以往100年中发展最为迅速的领域之一大批珍贵文物的出土不断刷新人们对文化中国的...         T_彭林   
..                                                 ...          ...   
701  p\r\n\tSTEM课程设计与案例分析课程它是br \r\n1国内领先针对STEM教育系统...        T_江丰光   
702  本课程作为数据科学的先导课和认知类课程致力于以形象生动的教学模式为学生普及数据挖掘大数据相关...         T_何隽   
703  p\r\n\t脱胎于清华经管学院与麦肯锡公司联合开设的线下金牌课程麦肯锡全球领导力在线课程共...        T_段志蓉   
704  物理化学上是一门运用物理学手段和技术研究化学问题的学科被誉为化学学科的大脑本课程将系统讲授物...        T_霍玉秋   
705  物理化学下是一门运用物理学手段和技术研究化学问题的学科被誉为化学学科的大脑本课程将系统讲授物...         T_郭迪   

          school                                           combined  
0      S_McGillX  自然灾害自主模式  无 p地球上没有一处地方不发生自然灾害当我们以科学的眼光看待这些自然灾害...  
1    S_TsinghuaX  2015年清华大学研究生学位论文答辩二  无先修要求 p学位论文答辩环节是研究生培养的重要环...  
2    S_TsinghuaX  2014年清华大学研究生学位论文答辩一  无先修要求 p学位论文答辩环节是研究生培养的重要环...  
3    S_TsinghuaX  2015年清华大学研究生学位论文答辩一  无先修要求 p学位论文答辩环节是研究生培养的重要环...  
4    S_TsinghuaX  文物精品与文化中国自主模式  无 p中国考古学是以往100年中发展最为迅速的领域之一大批珍贵...  
..           ...                                                ...  
701         S_TW  STEM课程设计与案例分析2019秋  无 p\r\n\tSTEM课程设计与案例分析课程它是...  
702  S_TsinghuaX  数据科学导论2019夏  本课程适合各类学科背景学生学习 本课程作为数据科学的先导课和认知类...  
703  S_TsinghuaX  麦肯锡全球领导力自主模式  Basicknowledgeofbusinessadminist...  
704        S_NEU  物理化学上2019夏  高等数学 物理化学上是一门运用物理学手段和技术研究化学问题的学科被誉...  
705        S_NEU  物理化学下2019夏  高等数学 物理化学下是一门运用物理学手段和技术研究化学问题的学科被誉...  

[706 rows x 7 columns]

# Modelling: CB

In [ ]:
vectorizer = TfidfVectorizer()
matrix = vectorizer.fit_transform(course_info_data["combined"])
cosine_similarities = linear_kernel(matrix,matrix)
course_id = course_info_data['id']
indices = pd.Series(course_info_data.index, index=course_info_data['id'])

In [ ]:
def content_recommender(title):
    idx = indices[title]
    sim_scores = list(enumerate(cosine_similarities[idx]))
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)
    sim_scores = sim_scores[1:6]
    course_indices = [i[0] for i in sim_scores]
    return course_id.iloc[course_indices]

In [ ]:
Y_pred = []

for i in X_test['course_9']:
  Y_pred.append(list(content_recommender(i)))

In [ ]:
Y_pred = pd.DataFrame(Y_pred)

In [ ]:
Y_pred

0    1    2    3    4
0     341  287  286  285  686
1     343  558  559  315  312
2     371  549  540  558  559
3     558  559  315  312  314
4     368  558  559  449  448
...   ...  ...  ...  ...  ...
1788  544   25  623  606  401
1789  375  390  449  448  364
1790  281  664   65   76  109
1791  279  677  214  698  373
1792  310  683  523  565  241

[1793 rows x 5 columns]

the top 5 predicted courses for all users in the test data.

# Evaluation - Actualness

The actualness evaluation represents how well the model predicts the final course sequence compared to the actual sequence. This evaluation is biased when the data quality is low. For example, people may take different subjects which causes the course sequence to become very random reducing the model's ability to predict the final course sequence.

In [ ]:
import math
from statistics import mean

def apk(actual, predicted, k=5):

    score = 0.0
    num_hits = 0.0

    for i,p in enumerate(predicted):

        if p == actual:
            num_hits += 1.0
            score += num_hits / (i+1.0)

    return score/k

CB_map_5_actual = []



for i in range(len(Y_test)):
  CB_map_5_actual.append(apk(Y_test.iloc[i],list(Y_pred.iloc[i]), k=5))

map_CB_actual = mean(CB_map_5_actual)
map_CB_actual

0.00831753113961703

In [ ]:
def hr(actual, predicted, k=5):

    num_hits = 0.0

    for i,p in enumerate(predicted):

        if p == actual:
            num_hits = 1.0
            pass

    return num_hits

CB_hit_5_actual = []

for i in range(len(Y_test)):
  CB_hit_5_actual.append(hr(Y_test.iloc[i],list(Y_pred.iloc[i]), k=5))

hr_CB_actual = mean(CB_hit_5_actual)
hr_CB_actual

0.06358059118795314

In [ ]:
def rr(actual, predicted, k=5):

    num_hits = 0.0
    score = 0.0

    for i,p in enumerate(predicted):

        if p == actual:
            num_hits = 1.0
            score = num_hits/(i+1)
            pass

    return score

CB_rr_5_actual = []

for i in range(len(Y_test)):
  CB_rr_5_actual.append(rr(Y_test.iloc[i],list(Y_pred.iloc[i]), k=5))

rr_CB_actual = mean(CB_rr_5_actual)
rr_CB_actual

0.041587655698085146

In [ ]:
def ndcg(actual, predicted, k=5):


    dcg = []
    idcg = []
    gain = 0.0

    for i in range(1,k+1):
      idcg.append(1/(math.log2(i+1)))

    for i,p in enumerate(predicted):

        if p == actual:
            gain = 1.0
            dcg.append(gain/(math.log2((i+1)+1)))

        else:
            dcg.append(0)

    return (sum(dcg))/(sum(idcg))

CB_ndcg_5_actual = []

for i in range(len(Y_test)):
  CB_ndcg_5_actual.append(ndcg(Y_test.iloc[i],list(Y_pred.iloc[i]), k=5))

ndcg_CB_actual = mean(CB_ndcg_5_actual)
ndcg_CB_actual

0.01595012690104388

In [ ]:
evaluation_actual_list = [[map_CB_actual, hr_CB_actual, rr_CB_actual, ndcg_CB_actual]]

In [ ]:
evaluation_actual = pd.DataFrame(evaluation_actual_list, columns = ['Mean Average Precision','Hit Ratio', 'Reciprocal Rank', "Normalized Discounted Cumulative Gain"])

In [ ]:
evaluation_actual

Mean Average Precision  Hit Ratio  Reciprocal Rank  \
0                0.008318   0.063581         0.041588   

   Normalized Discounted Cumulative Gain  
0                                0.01595

# Evaluation - Relevancy

On the other hand, the relevancy evaluation represents how well the model predicts the final course sequence that is relevant to the student based on actual sequences, schools, teachers, and prerequisite concepts to be relevant. This evaluation is added due to the actualness evaluation may give irrelevant courses to the student.

In [ ]:
import math
from statistics import mean
def apk(x_actual,actual, predicted, k=5):

    score = 0.0
    num_hits = 0.0

    for i,p in enumerate(predicted):

        p_c = course_decode[p]
        actual_c = course_decode[actual]
        actual_x = course_decode[x_actual]

        p_s = dict_lookup_course_school[p_c]
        actual_s = dict_lookup_course_school[actual_c]

        p_t = dict_lookup_course_school[p_c]
        actual_t = dict_lookup_course_school[actual_c]

        p_r = dict_lookup_course_prerequisite[p_c]
        actual_r = dict_lookup_course_concept[actual_x]


        if p == actual:
            num_hits += 1.0
            score += num_hits / (i+1.0)

        elif p_s == actual_s and p_t == actual_t and len(list(set(p_r) & set(actual_r))) >= 5:
            num_hits += 1.0
            score += num_hits / (i+1.0)

    return score/k

CB_map_5 = []



for i in range(len(Y_test)):
  CB_map_5.append(apk(X_test['course_9'].iloc[i],Y_test.iloc[i],list(Y_pred.iloc[i]), k=5))

map_CB = mean(CB_map_5)
map_CB

0.19927309908905

In [ ]:
def hr(x_actual, actual, predicted, k=5):

    num_hits = 0.0

    for i,p in enumerate(predicted):

      p_c = course_decode[p]
      actual_c = course_decode[actual]
      actual_x = course_decode[x_actual]

      p_s = dict_lookup_course_school[p_c]
      actual_s = dict_lookup_course_school[actual_c]

      p_t = dict_lookup_course_school[p_c]
      actual_t = dict_lookup_course_school[actual_c]

      p_r = dict_lookup_course_prerequisite[p_c]
      actual_r = dict_lookup_course_concept[actual_x]

      if p == actual:
          num_hits = 1.0
          pass

      elif p_s == actual_s and p_t == actual_t and len(list(set(p_r) & set(actual_r))) >= 5:
          num_hits = 1.0
          pass

    return num_hits

CB_hit_5 = []

for i in range(len(Y_test)):
  CB_hit_5.append(hr(X_test['course_9'].iloc[i],Y_test.iloc[i],list(Y_pred.iloc[i]), k=5))

hr_CB = mean(CB_hit_5)
hr_CB

0.3714445064138316

In [ ]:
def rr(x_actual, actual, predicted, k=5):

    num_hits = 0.0
    score = 0.0

    for i,p in enumerate(predicted):

      p_c = course_decode[p]
      actual_c = course_decode[actual]
      actual_x = course_decode[x_actual]

      p_s = dict_lookup_course_school[p_c]
      actual_s = dict_lookup_course_school[actual_c]

      p_t = dict_lookup_course_school[p_c]
      actual_t = dict_lookup_course_school[actual_c]

      p_r = dict_lookup_course_prerequisite[p_c]
      actual_r = dict_lookup_course_concept[actual_x]

      if p == actual:
          num_hits = 1.0
          score = num_hits/(i+1)
          pass

      elif p_s == actual_s and p_t == actual_t and len(list(set(p_r) & set(actual_r))) >= 5:
          num_hits = 1.0
          score = num_hits/(i+1)
          pass

    return score
CB_rr_5 = []

for i in range(len(Y_test)):
  CB_rr_5.append(rr(X_test['course_9'].iloc[i],Y_test.iloc[i],list(Y_pred.iloc[i]), k=5))

rr_CB = mean(CB_rr_5)
rr_CB

0.1284067670570738

In [ ]:
def ndcg(x_actual,actual, predicted, k=5):


    dcg = []
    idcg = []
    gain = 0.0

    for i in range(1,k+1):
      idcg.append(1/(math.log2(i+1)))

    for i,p in enumerate(predicted):

        p_c = course_decode[p]
        actual_c = course_decode[actual]
        actual_x = course_decode[x_actual]

        p_s = dict_lookup_course_school[p_c]
        actual_s = dict_lookup_course_school[actual_c]

        p_t = dict_lookup_course_school[p_c]
        actual_t = dict_lookup_course_school[actual_c]

        p_r = dict_lookup_course_prerequisite[p_c]
        actual_r = dict_lookup_course_concept[actual_x]


        if p == actual:
            gain = 1.0
            dcg.append(gain/(math.log2((i+1)+1)))


        elif p_s == actual_s and p_t == actual_t and len(list(set(p_r) & set(actual_r))) >= 5:
            gain = 1.0
            dcg.append(gain/(math.log2((i+1)+1)))

        else:
            dcg.append(0)

    return (sum(dcg))/(sum(idcg))

CB_ndcg_5 = []

for i in range(len(Y_test)):
  CB_ndcg_5.append(ndcg(X_test['course_9'].iloc[i],Y_test.iloc[i],list(Y_pred.iloc[i]), k=5))

ndcg_CB = mean(CB_ndcg_5)
ndcg_CB

0.23452502763914188

In [ ]:
evaluation_relevant_list = [[map_CB, hr_CB, rr_CB, ndcg_CB]]

In [ ]:
evaluation_relevant = pd.DataFrame(evaluation_relevant_list, columns = ['Mean Average Precision','Hit Ratio', 'Reciprocal Rank', "Normalized Discounted Cumulative Gain"])

In [ ]:
evaluation_relevant

Mean Average Precision  Hit Ratio  Reciprocal Rank  \
0                0.199273   0.371445         0.128407   

   Normalized Discounted Cumulative Gain  
0                               0.234525